---
title: "B.1.251 sequences and S-gene positive pillar 2 cases"
author: "Rob Challen"
date: '`r format(Sys.Date(), "%d-%m-%Y")`'
output: 
  pdf_document :
    fig_caption: yes
header-includes:
  \usepackage{float}
  \floatplacement{figure}{H}
knit: (function(inputFile, encoding,...) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "~/Dropbox/covid19/sa-variant/")})
fig_width: 7
fig_height: 5
out.width: "100%"
---


```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo = FALSE,
  warning = FALSE,
  message = FALSE
)

devtools::load_all("~/Git/standard-print-output/")
devtools::load_all("~/Git/uk-covid-datatools/")

library(tidyverse)
library(patchwork)
library(rgdal)
library(ggplot2)
library(ggspatial)
library(rgeos)
library(maptools)
library(patchwork)
library(sp)
library(sf)

ggplot2::theme_set(standardPrintOutput::defaultFigureLayout())
standardPrintOutput::setDefaults()
ukcovidtools::setup()
ukcovidtools::reload()
```

```{r eval=FALSE}
devtools::load_all("~/Git/uk-covid-datatools/")
ukcovidtools::reload()

setwd("~/Git/uk-covid-datatools/")
if (!exists("combinedSpositives")) source("./vignettes/s-pos-voc/s-pos-data.R")

# Define areas of concern ---
# LAD codes in 3 regions


```

# Descriptive stats 

```{r}
# quick count summary ----

aprilCases = combinedCases %>% filter(earliest_specimen_date > "2021-04-01")
with(aprilCases, table(sGene,type,useNA = "ifany"))

genomics %>% 
  group_by(variant) %>% 
  summarise(
    count = n(), 
    earliest = min(specimen_date,na.rm = TRUE), 
    latest = max(specimen_date,na.rm = TRUE)
  ) %>%
  mapVOCtoLineage(vocVar = variant) %>% 
  readr::write_csv(paste0("~/Dropbox/covid19/sa-variant/CTAS-genomics-summary-",Sys.Date(),".csv"))

# TODO - more required here
```

## Pillar 2 S-gene tests count and TaqPath coverage by LTLA 

```{r}

matched = ll %>% filter(pillar == "Pillar 2") %>% inner_join(sgll, by="FINALID", suffix = c("",".sgene")) %>% filter(specimen_date >= specimen_date.sgene & specimen_date <= specimen_date.sgene + 4 ) %>% select(FINALID)

tmp = bind_rows(
  ll %>% filter(pillar == "Pillar 2") %>% semi_join(matched, by="FINALID") %>% mutate(sgeneKnown=TRUE),
  ll %>% filter(pillar == "Pillar 2") %>% anti_join(matched, by="FINALID") %>% mutate(sgeneKnown=FALSE)
) %>% filter(specimen_date > earliestDate & specimen_date < lastDate)

tqPathCoverage = tmp %>% group_by(LTLA_code, sgeneKnown) %>% summarise(count = n()) %>%
  group_by(LTLA_code) %>%
  mutate(binom::binom.confint(count, sum(count), conf.level = 0.95, methods = "wilson")) %>%
  mutate(coverage = sprintf("%1.0f%%", mean*100)) %>% 
  filter(sgeneKnown==TRUE) %>%
  select(-sgeneKnown)

tqPathCoverage %>% readr::write_csv(paste0("~/Dropbox/covid19/sa-variant/taq-path-coverage-",earliestDate,"-to-",lastDate,"-by-ltla.csv" ))

map = dpc$geog$getMap("LAD19")
map2 = map %>% left_join(tqPathCoverage, by=c("code"="LTLA_code")) %>% 
  mutate(
    mean = ifelse(!is.finite(mean),0,mean),
    n = ifelse(is.na(n),0,n)
    ) %>% filter(code %>% stringr::str_starts("E"))

(ggplot(map2,aes(fill=mean*100))+geom_sf(size=0.3)+
    scale_fill_gradient(low="#FFFFFF",high="#0080FF",lim=c(0,100),name="% coverage")+
    labs(subtitle = paste0(earliestDate," to ",lastDate))
) %>% standardPrintOutput::saveFigure(maxWidth = 6,maxHeight = 6,filename = "~/Dropbox/covid19/sa-variant/LTLA-taqpath-coverage")

(ggplot(map2,aes(fill=n))+geom_sf(size=0.3)+
    scale_fill_gradient(low="#FFFFFF",high="#FF8000",name="Pillar 2 tests")+
    labs(subtitle = paste0(earliestDate," to ",lastDate))
) %>% standardPrintOutput::saveFigure(maxWidth = 6,maxHeight = 6,filename = "~/Dropbox/covid19/sa-variant/LTLA-cumulative-cases")


```

# LTLA data and plots

```{r}

convertDist = function(delay, ...) {
  combinedCases %>% 
  filter(earliest_specimen_date > earliestDate-delay & earliest_specimen_date < lastDate-delay) %>% mutate(
    category = case_when(
      sequenced & type == "B.1.617.2" ~ "B.1.617.2",
      sequenced & type == "B.1.1.7" ~ "B.1.1.7",
      sequenced ~ "other VOC/VUI",
      !sequenced & sGene == "positive" ~ "unsequenced S+",
      !sequenced & sGene == "negative" ~ "unsequenced S-",
      TRUE ~ "unknown"
    ) %>% forcats::as_factor() %>% forcats::fct_relevel("unsequenced S+","B.1.617.2","unsequenced S-","B.1.1.7","other VOC/VUI","unknown")
  ) %>% 
  group_by(code = LTLA_code, name=LTLA_name, category) %>%
  summarise(count= n()) %>%
  # mutate(delay = delay) %>%
  ungroup() 
}

ltlaCounts = convertDist(0) %>% left_join( convertDist(7) %>% rename(count7 = count) ) %>% mutate(change = (count/count7-1)*100)


ltlaCounts2 = combinedCases %>% 
  filter(earliest_specimen_date > earliestDate-7 & earliest_specimen_date < lastDate-7) %>% mutate(
    category = case_when(
      sequenced & type == "B.1.617.2" ~ "B.1.617.2",
      sequenced & type == "B.1.1.7" ~ "B.1.1.7",
      sequenced ~ "other VOC/VUI",
      !sequenced & sGene == "positive" ~ "unsequenced S+",
      !sequenced & sGene == "negative" ~ "unsequenced S-",
      TRUE ~ "unknown"
    ) %>% forcats::as_factor() %>% forcats::fct_relevel("unsequenced S+","B.1.617.2","unsequenced S-","B.1.1.7","other VOC/VUI","unknown")
  ) %>% 
  group_by(code = LTLA_code, name=LTLA_name, category) %>%
  summarise(count= n()) %>%
  ungroup() 

```


```{r}
ifna = function(x,y) ifelse(is.na(x),y,x)

plotWithLabels = function(ltlaCounts, NHSERname = NULL, labelsPerPlot = 4, nrow=2) {
  
  
  ltlaCounts = ltlaCounts %>% filter(category %in% c("unsequenced S+","B.1.617.2","B.1.1.7"))
  ladMap = dpc$geog$getMap("LAD19")
  
  if(!is.null(NHSERname)) {
  
    ltlaLists = dpc$codes$getTransitiveClosure() %>% 
      filter(fromCodeType=="LAD",toCodeType=="NHSER") %>% 
      dpc$codes$findNamesByCode(codeVar = toCode) %>%
      filter(name == NHSERname) %>% pull(fromCode)
  
    
  } else {
    ltlaLists = ladMap %>% filter(code %>% stringr::str_starts("E")) %>% pull(code)
  }
  
  ladMap = ladMap %>% filter(code %in% ltlaLists)
  
  #aocMap2 = aocMap %>% filter(code %in% ltlaLists)
  
  ltlas = ladMap %>% as_tibble() %>% select(code,name) %>% distinct() %>%
    mutate(label = code %>% stringr::str_replace("E0(.)0+(.+)","\\1.\\2"))
  categories = ltlaCounts %>% select(category) %>% distinct() %>% arrange(category)

  ltlaCounts2 = ltlas %>% select(-name) %>% inner_join(categories, by=character()) %>%
    left_join(ltlaCounts %>% select(-name), by=c("code","category")) %>% 
    mutate(count = ifelse(is.na(count),0,count))
  
  mapVoc = ltlaCounts2 %>% left_join(ladMap, by=c("code"), suffix=c("",".map")) %>% sf::st_as_sf()
  
  labels = mapVoc %>% 
    sf::st_centroid() %>%
    mutate(x=sf::st_coordinates(.)[,"X"], y=sf::st_coordinates(.)[,"Y"]) %>% 
    filter(code %in% ltlaLists) %>%
    as_tibble() %>%
    group_by(category) %>% 
    arrange(desc(count)) %>% 
    filter(row_number() <= labelsPerPlot) %>%
    ungroup()
  
  p1 = lapply(categories$category, function(cat) {
    breaks = mapVoc %>% filter(category==cat & count>0) %>% pull(count) %>% quantile(c(0.5,0.75,0.9,0.95,0.975))
    breaks = sort(unique(round(c(5,unname(breaks[breaks>5])))))
    plotlabels = c("0",ifelse(c(0,breaks)+1 == c(breaks,Inf), as.character(breaks) , paste0( as.character(c(0,breaks)+1),c(rep("-",length(breaks)),"+"),c(as.character(breaks),""))))
    tmp = mapVoc %>% filter(category==cat) %>% mutate(count2=cut(count,breaks = c(-Inf,0,breaks,Inf),labels = plotlabels,include.lowest = FALSE))
    ggplot()+
    #geom_sf(data=dpc$geog$getMap("LAD19") ,size=0.05)+
      geom_sf(data=tmp,mapping=aes(fill=count2),size=0.1,colour="white")+
      scale_fill_viridis_d(option = "inferno", name="")+
      standardPrintOutput::defaultMapLayout()+
      standardPrintOutput::narrower()+
      standardPrintOutput::smallLegend(textSize = 6)+ #spaceLegend = 0.8)+
      facet_wrap(vars(category))+
      ggrepel::geom_label_repel(
        data = labels %>% filter(category==cat), 
        aes(x=x,y=y,label=label),min.segment.length = 0,segment.colour = "cyan",fill="#80808080",colour="cyan",size=1.5)
  })
  
  
  #print(labs %>% mutate(l = paste0(label,": ",name)) %>% pull(l) %>% paste0(collapse = "; "))
  p4 = standardPrintOutput::simpleFigureTable(labels %>% as_tibble() %>% select(label,name) %>% distinct() %>% arrange(label),pts = 5,unwrapped = TRUE)
  #browser()
  return(patchwork::wrap_plots(c(p1,list(p4)),nrow=nrow)+patchwork::plot_annotation(title = paste0("cases between ",earliestDate," and ",lastDate))) #, widths = c(3,3,3,2)))#+patchwork::plot_layout(ncol = 2))
}
```

# VOC England maps

```{r}
plotWithLabels(
  ltlaCounts
) %>% standardPrintOutput::saveTwoThirdPageFigure(filename="~/Dropbox/covid19/sa-variant/SPosVsSequencesEngland")

plotWithLabels(
  ltlaCounts, nrow=1
) %>% standardPrintOutput::saveFigure(maxWidth=10,maxHeight=5,filename="~/Dropbox/covid19/sa-variant/SPosVsSequencesEnglandPresentation")

plotWithLabels(
  ltlaCounts,
  labelsPerPlot = 8
) %>% standardPrintOutput::saveFullPageFigure(filename="~/Dropbox/covid19/sa-variant/SPosVsSequencesEnglandLSCAPE")



```



# VOC local maps
```{r}

# TODO: change to use list of LTLAs rather than coordinates
regions = dpc$codes$getCodes() %>% filter(codeType == "NHSER" & status=="live")


# locations = list(
#   `Leicester and Nottingham` = c(52.5,-1.6,53,-0.9),
#   `Bolton, Preston & Blackburn` = c(53.5,-3.2,54,-2),
#   
#   #`Leeds, Bradford, Huddersfield` = c(53.5,-2.1,54,-1),
#   #`Manchester, Liverpool` = c(52.8,-3.2,53.75,-1.8),
#   #`Newcastle, Sunderland` = c(54.5,-2.15,55.4,-0.7),
#   #`Southampton, Portsmouth` = c(50.4,-2,51.4,0.2),
#   `Birmingham` = c(52.1,-2.7,52.9,-1.15),
#   #`M4 corridor` = c(51,-2.8,51.9,-0.8),
#   `London` = c(51.15,-0.8,51.85,0.6)
#   
# )

for(name in regions$name) {
  plotWithLabels(ltlaCounts = ltlaCounts, NHSERname = name, labelsPerPlot = 4) %>% 
    standardPrintOutput::saveTwoThirdPageFigure(filename=paste0("~/Dropbox/covid19/sa-variant/SPosVsSequences - ",name))
}

```


# LTLA VOC count table(s)

```{r}

aocDf %>% rename(name = ltlaName) %>% select(area,code,name,population) %>% mutate(population = sprintf("%1.0f",population)) %>% group_by(area) %>% standardPrintOutput::saveTable(filename="~/Dropbox/covid19/sa-variant/areasOfConcernTable")

```





```{r}



tableCutoff = function() {
  
  tmpTq = tqPathCoverage %>% inner_join(aocDf, by=c("LTLA_code"="code")) %>% 
    group_by(area) %>% summarise(x= sum(x), n=sum(n)) %>% mutate(binom::binom.confint(x=x,n=n,methods = "wilson")) %>%
    mutate(coverage = sprintf("%1.0f%%",mean*100)) %>% select(area,coverage)
  
  change = ltlaCounts %>% mutate(
    compat = case_when(
      category %in% c("unsequenced S+","B.1.617.2") ~ "B.1.617.2 compatible", 
      category %in% c("unsequenced S-","B.1.1.7") ~ "B.1.1.7 compatible",
      TRUE ~ "other"
    )) %>% left_join(aocDf, by="code") %>%
    mutate(area = ifelse(is.na(area),"other",as.character(area))) %>%
    group_by(area, compat) %>%
    summarise(count = sum(count,na.rm = TRUE), count7 = sum(count7,na.rm = TRUE)) %>% 
    mutate(change = count/count7-1) %>%
    mutate(entry = sprintf("%1.0f%%", change*100)) %>%
    select(-count7, -count,-change) %>%
    pivot_wider(names_from = compat,values_from=entry)
    
  clustDf = tibble(group = names(clusters) %>% ordered(names(clusters)), area = clusters) %>% unnest(cols=area)
  # expected = ltlaCounts %>% filter(category=="unsequenced S+") %>% inner_join(
  #     tqPathCoverage %>% select(LTLA_code,coverage,meanCov = mean,lowCov = lower,hiCov = upper), by=c("code"="LTLA_code")
  #   ) %>%
  #   mutate(`expected S+` = sprintf("%1.0f\u2013%1.0f",count/hiCov,count/lowCov),bestCase=count/hiCov) %>% select(code,coverage,`expected S+`,bestCase)
  # 
  # browser()
  
  table = ltlaCounts %>% 
    #group_by(name, code, change) %>% 
    #mutate(label = code %>% stringr::str_replace("E0(.)0+(.+)","\\1.\\2")) %>%
    ungroup() %>% 
    left_join(aocDf, by="code") %>%
    mutate(area = ifelse(is.na(area),"other",as.character(area))) %>%
    group_by(area,category) %>% 
    summarise(
      count = sum(count,na.rm = TRUE),
    ) %>% 
    group_by(area) %>%
    mutate(tmp = sum(ifelse(category %in% c("unsequenced S+","B.1.617.2"),count,0))) %>%
    mutate(entry = ifelse(count>0 & count<5,"<5",sprintf("%1.0f", count))) %>%
    #filter(tmp > cutoff) %>%
    select(-count,-tmp) %>%
    pivot_wider(names_from = category, values_from=entry,values_fill="-") %>%
    inner_join(tmpTq, by=c("area")) %>%
    inner_join(change, by=c("area")) %>%
    inner_join(clustDf, by=c("area")) %>%
    filter(`B.1.617.2` != "-") %>% 
    select(group,area,`B.1.617.2`,`unsequenced S+`,`B.1.617.2/S+ %/week`=`B.1.617.2 compatible`,`B.1.1.7`,`unsequenced S-`,`B.1.1.7/S- %/week` =`B.1.1.7 compatible`,`other VOC/VUI`,`S-gene unknown`=`unknown`,`TaqPath coverage`=coverage) %>% as_tibble()
  return(table)
}

tableCutoff() %>% group_by(group) %>% standardPrintOutput::saveTable("~/Dropbox/covid19/sa-variant/summaryCounts")
#tableCutoff(90) %>% group_by(area) %>% standardPrintOutput::saveTable("~/Dropbox/covid19/sa-variant/summaryCountsShort")
```

# Area of concern epi curves

```{r}




# inboundTrav = readr::read_csv("~/Dropbox/covid19/sa-variant/Inputs/20210427_inbound_positive_cases_by_LTLA.csv")
# 
# trav2 = aocDf %>% inner_join(inboundTrav, by=c("code"="LTLA_code")) %>% group_by(area,date) %>% summarise(count = sum(imported_cases))
# 
# p5 = (ggplot(trav2 %>% filter(date>earliestDate),aes(x=date,y=count))+
#     geom_bar(stat="identity")+
#     facet_wrap(vars(area),nrow=1)+
#     coord_cartesian(xlim=c(earliestDate,lastDate))+
#     standardPrintOutput::smallLegend()+geom_vline(xintercept = max(trav2$date), colour="blue")) #%>% standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/AOCImportation")

doEpiCurve = function(combinedData, areas = aocDf, earliestDate = as.Date("2021-04-01")) {
  tmp = combinedData %>% inner_join(areas, by=c("LTLA_code"="code")) %>% mutate(
    context = case_when(
      asymptomatic_indicator == "N" ~ "symptoms",
      asymptomatic_indicator == "Y" ~ "screening",
      asymptomatic_indicator == "U" ~ "unknown",
      TRUE ~ NA_character_
    ) %>% ordered(c("symptoms","screening","unknown")),
    ageCat = case_when(
      age < 6 ~ "<6",
      age < 19 ~ "6-18",
      age < 22 ~ "19-21",
      age < 45 ~ "22-44",
      age < 65 ~ "45-64",
      age < 80 ~ "65-79",
      TRUE ~ "80+"
    ) %>% ordered(c("<6","6-18","19-21","22-44","45-64","65-79","80+"))
  ) %>% rename(date = earliest_specimen_date)
  
  p1 = (ggplot(tmp %>% filter(date>earliestDate & date<=lastDate),aes(x=date,fill=type))+
      geom_histogram(binwidth = 1)+
      facet_wrap(vars(area),nrow=1)+
      coord_cartesian(xlim=c(earliestDate,lastDate))+
      scale_fill_brewer(palette = "Set1",na.value = "grey70",name=NULL)+
      standardPrintOutput::smallLegend()) #%>% standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/AOCByType")
  
  p3 = (ggplot(tmp %>% filter(date>earliestDate & date<=lastDate),aes(x=date,fill=exposure_type))+
      geom_histogram(binwidth = 1)+
      facet_wrap(vars(area),nrow=1)+
      coord_cartesian(xlim=c(earliestDate,lastDate))+
      scale_fill_brewer(palette = "Paired",na.value = "grey70",name=NULL)+
      standardPrintOutput::smallLegend()) #%>% standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/AOCByTravel")
  
  # (ggplot(tmp %>% filter(date>earliestDate),aes(x=date,fill=reinfect56))+
  #     geom_histogram(binwidth = 1)+
  #     facet_wrap(vars(area),ncol=1)+
  #     standardPrintOutput::smallLegend()) %>% standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/AOCByReinfection")
  
  p2 = (ggplot(tmp %>% filter(date>earliestDate & date<=lastDate),aes(x=date,fill=context))+
      geom_histogram(binwidth = 1)+
      facet_wrap(vars(area),nrow=1)+
      coord_cartesian(xlim=c(earliestDate,lastDate))+
      scale_fill_brewer(palette = "Dark2",na.value = "grey70",name=NULL)+
      standardPrintOutput::smallLegend()) #%>% standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/AOCByAsympInd")
  
  p4 = (ggplot(tmp %>% filter(date>earliestDate & date<=lastDate),aes(x=date,fill=ethnicity_final))+
      geom_histogram(binwidth = 1)+
      facet_wrap(vars(area),nrow=1)+
      coord_cartesian(xlim=c(earliestDate,lastDate))+
      scale_fill_brewer(palette = "Set2",na.value = "grey70",name=NULL)+
      standardPrintOutput::smallLegend()) #%>% standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/AOCByEthnic")
  
  p5 = (ggplot(tmp %>% filter(date>earliestDate & date<=lastDate),aes(x=date,fill=ageCat))+
      geom_histogram(binwidth = 1)+
      facet_wrap(vars(area),nrow=1)+
      coord_cartesian(xlim=c(earliestDate,lastDate))+
      scale_fill_brewer(palette = "Accent",na.value = "grey70",name=NULL)+
      standardPrintOutput::smallLegend()) #%>% standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/AOCByEthnic")
  
  
  # (ggplot(tmp %>% filter(date>earliestDate & date<=lastDate),aes(x=date,fill=isolation_cat))+
  #     geom_histogram(binwidth = 1)+
  #     facet_wrap(vars(area),nrow=1)+
  #     coord_cartesian(xlim=c(earliestDate,lastDate))+
  #     standardPrintOutput::smallLegend()+
  #     scale_fill_brewer(palette = "Dark2",na.value = "grey70"))
  
    #%>% standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/AOCByEthnic")
  
  # (ggplot(tmp %>% filter(date>earliestDate),aes(x=date,fill=ctas_status))+
  #     geom_histogram(binwidth = 1)+
  #     facet_wrap(vars(area),ncol=1)+
  #     standardPrintOutput::smallLegend()) %>% standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/AOCByCTAS")
  
  (
  (p1+standardPrintOutput::hideX()+ylab("S+ counts")+theme(legend.justification='left'))+
  (p2+standardPrintOutput::hideX()+ylab("S+ counts")+theme(legend.justification='left'))+
  (p3+standardPrintOutput::hideX()+ylab("S+ counts")+theme(legend.justification='left'))+
  (p4+standardPrintOutput::hideX()+ylab("S+ counts")+theme(legend.justification='left'))+
  (p5+ylab("S+ counts")+theme(legend.justification='left'))+
  patchwork::plot_layout(ncol=1)) %>% return()
}

clusters2 = list(
  "NHS Regions" = c("North West","Midlands","London"),
  "NW Cluster" = c("Bolton","M65 Corridor","Manchester"), #"M6 Corridor","Sefton & Lpl"
  "Midlands" = c("Nottingham","Leicester","Birmingham") #"Bedford etc",
)

lapply(names(clusters2), function(name) {
  doEpiCurve(combinedSpositives, aocDf %>% filter(area %in% clusters2[[name]])) %>% standardPrintOutput::saveFigure(paste0("~/Dropbox/covid19/sa-variant/AOCMultiplot-",name,"-",Sys.Date()),maxWidth = 8,maxHeight = 6)
})

#doEpiCurve(combinedSpositives, aocDf %>% filter(area %in% c("Bolton","M65 Corridor","M6 Corridor","Sefton & Lpl","Manchester"))) %>% standardPrintOutput::saveFigure("~/Dropbox/covid19/sa-variant/AOCMultiplotNWCluster",maxWidth = 10,maxHeight = 6)
#doEpiCurve(combinedSpositives, aocDf %>% filter(area %in% c("Bedford etc","Nottingham","Leicester","Birmingham"))) %>% standardPrintOutput::saveFigure("~/Dropbox/covid19/sa-variant/AOCMultiplotMidlands",maxWidth = 10,maxHeight = 6)

#doEpiCurve(combinedSpositives, aocDf %>% filter(area != "NW Cluster")) %>% standardPrintOutput::saveFigure("~/Dropbox/covid19/sa-variant/AOCMultiplotNotNWLSCPE",maxWidth = 10,maxHeight = 6)
#doEpiCurve(combinedSnotNegatives) %>% standardPrintOutput::saveFigure("~/Dropbox/covid19/sa-variant/AOCMultiplotNotNegLSCPE",maxWidth = 10,maxHeight = 6)

```


# Figure 1 - proportions and counts

```{r}
vocLevels = c("B.1.1.7","B.1.351","B.1.617.1","B.1.617.2","P.1 & P.2","other VOC/VUI","non VOC/VUI")
palVoc = scales::brewer_pal(palette="Set2")(7)
names(palVoc) = vocLevels

pSposIsVariant = allVoc %>%
  mutate(date = as.Date(ifelse(is.na(specimen_date),episode_date,specimen_date),"1970-01-01")) %>%
  #inner_join(sgll %>% select(FINALID, sgtf_under30CT), by=c("finalid"="FINALID")) %>%
  filter(sGene == "positive") %>%
  mapVOCtoLineage(vocVar=variant,lineages = c("B.1.617.2","B.1.617.1","B.1.351","P.1 & P.2")) %>%
  # mutate(
  #   detected = ifelse(asymptomatic_indicator == "Y","screening & LFT","symptoms")
  # ) %>%
  group_by(type, date) %>%
  summarise(count= n()) %>% 
  ungroup() %>%
  tidyr::complete(type,date,fill=list(count=0)) %>%
  group_by(type) %>% 
  arrange(date) %>%
  mutate(Roll.count = stats::filter(count,filter=rep(1,14)/14,sides=1)) %>%
  filter(!is.na(Roll.count)) %>% 
  group_by(date) %>% 
  mutate(binom::binom.confint(Roll.count, sum(Roll.count), conf.level = 0.95, methods = "wilson")) %>% 
  ungroup() %>%
  mutate(type = type %>% ordered(vocLevels))


#(countSposSeqByType+standardPrintOutput::narrower()) %>% standardPrintOutput::saveSixthPageFigure(filename="~/Dropbox/covid19/sa-variant/countSposSeqByType")
```

```{r}

countSposSeqByType = ggplot(pSposIsVariant %>% filter(date > "2021-01-21"), aes(x=date, y=Roll.count,colour=type))+
  #geom_smooth(se = FALSE)+
  geom_line()+
  standardPrintOutput::smallLegend(textSize = 8)+
  ylab("Number of sequenced S+")+
  theme(legend.title = element_blank())+
  scale_y_continuous(trans="log1p",breaks = c(1,2,5,10,20,50,100,200,500,1000,2000,5000),limits=c(0,NA))+
  scale_colour_manual(values = palVoc,aesthetics = c("colour","fill"), name="Lineage", drop=FALSE)+
  scale_x_date(date_breaks = "2 weeks", date_labels = "%d-%m")


pSposPlot = ggplot(pSposIsVariant %>% filter(date > "2021-01-21"), aes(x=date, y=mean, ymin = lower,ymax=upper,colour=type, fill=type))+
  #geom_ribbon(alpha=0.2,colour=NA)+
  #geom_line(position="stack")+
  geom_area(alpha=0.25,outline.type="both")+
  standardPrintOutput::smallLegend(textSize = 8)+
  ylab("Proportion of sequenced S+")+
  #theme(legend.title = element_blank())+
  scale_fill_manual(values = palVoc,aesthetics = c("colour","fill"),name=NULL)+
  scale_x_date(date_breaks = "2 weeks", date_labels = "%d-%m")

# pSposPlot %>% standardPrintOutput::saveSixthPageFigure(filename="~/Dropbox/covid19/sa-variant/propSPosIsVOC")
# 
# rightPanel = ((countSposSeqByType+guides(colour=guide_none(),fill=guide_none())+standardPrintOutput::narrower()+standardPrintOutput::hideX())+
#     (pSposPlot+standardPrintOutput::narrower()+theme(legend.justification='left'))+
#     patchwork::guide_area()+
#     patchwork::plot_annotation(tag_levels = "A")+
#     patchwork::plot_layout(ncol=1,guides = "collect",heights = c(3,3,1))) 

# %>% standardPrintOutput::saveFigure(maxWidth=3,maxHeight = 6, filename="~/Dropbox/covid19/sa-variant/Fig2ProportionsSequences")


```

```{r}

lsoaCentroid = dpc$geog$getMap("LSOA11")
lsoaCentroid = lsoaCentroid %>% sf::st_centroid()
mapping = dpc$geog$getContainedIn(inputSf = lsoaCentroid,outputShape = aocMap,outputIdVar = areaName)
modelled = readr::read_csv("~/Dropbox/covid19/sa-variant/Inputs/ByLSOAVariantRateswDGP.csv/ByLSOAVariantRateswDGP.csv")

modelledEng = modelled %>% 
  filter(LSOA_code %>% stringr::str_starts("E")) %>% 
  group_by(week, type) %>% 
  summarise(variantGR = sum(VariantRate,na.rm = TRUE)) %>% 
  mutate(area="England", variantDaily = variantGR/7, date = week+3.5) %>%
  mutate(type = type %>% ordered(vocLevels))



```


```{r}
seqBySgeneCounts = allVoc %>%
  mutate(date = as.Date(ifelse(is.na(specimen_date),episode_date,specimen_date),"1970-01-01")) %>%
  group_by(sGene, date) %>%
  summarise(count= n()) %>% arrange(date) %>%
  mutate(
    Roll.count = stats::filter(count,filter=rep(1,14)/14)
  )
  


# seqCountsBySgene = ggplot(seqBySgeneCounts %>% filter(date > "2020-11-01"),aes(x=date,colour=sGene))+
#   geom_point(mapping=aes(y=count),size=0.5)+
#   geom_line(mapping=aes(y=Roll.count))+
#   scale_y_continuous(trans="log1p",breaks = c(1,2,5,10,20,50,100,200,500,1000,2000,5000),limits=c(0,NA))+standardPrintOutput::narrower()+
#   ylab("count of COG-UK sequencing")

sgeneCounts = allEpisodes %>%
  #filter(sgtf_under30ct == 0) %>%
  group_by(sGene,date=earliest_specimen_date) %>%
  summarise(count= n()) %>% arrange(date) %>%
  mutate(
    Roll.count = stats::filter(count,filter=rep(1,14)/14)
  ) 

countsBySgene2 = 
  ggplot(
    bind_rows(
      sgeneCounts %>% mutate(source = "Pillar 2 positives"),
      seqBySgeneCounts %>% mutate(source = "Sequencing")
    ) %>% filter(sGene %in% c("positive","negative","unknown")),
    aes(x=date,y=count, linetype=source, colour=sGene)
  )+
  geom_line(aes(y=Roll.count))+
  geom_point(aes(y=count,shape=source),size=0.5)+
  #geom_point(data=modelledEng, mapping = aes(x=date,y=variantDaily,fill=type),inherit.aes = FALSE,shape=21,stroke=0)+
  #scale_fill_manual(values = palVoc, name="Lineage", drop=FALSE)+
  #geom_smooth(aes(y=count))+
  scale_y_continuous(trans="log1p",breaks = c(1,2,5,10,20,50,100,200,500,1000,2000,5000),limits=c(0,NA))+
  scale_linetype(name="Source")+
  scale_colour_discrete(name="S gene status")+
  #standardPrintOutput::narrower()+
  standardPrintOutput::smallLegend(textSize = 8,spaceLegend = 1)+
  ylab("Number of cases")+
  scale_x_date(limits = c(as.Date("2021-01-21"),NA), date_breaks = "2 weeks", date_labels = "%d-%m")


  combinedPlot2 = ((countsBySgene2+theme(legend.justification='left',axis.text.x = element_text(angle=60))+standardPrintOutput::smallLegend(spaceLegend = 1)) |
    ( (countSposSeqByType+guides(colour=guide_none(),fill=guide_none())+standardPrintOutput::narrower()+standardPrintOutput::hideX()) /
    (pSposPlot+guides(colour=guide_none(),fill=guide_none())+standardPrintOutput::narrower()+theme(axis.text.x = element_text(angle=60)))))+
    patchwork::plot_annotation(tag_levels = "A")+patchwork::plot_layout(guides="collect")
  
  
  countsBySgene = 
    ggplot(
      bind_rows(
        sgeneCounts %>% mutate(source = "Pillar 2 positives"),
        seqBySgeneCounts %>% mutate(source = "Sequencing")
      ) %>% filter(sGene %in% c("positive","negative")),
      aes(x=date,y=count, linetype=source, colour=sGene)
    )+
    geom_line(aes(y=Roll.count))+
    geom_point(aes(y=count,shape=source),size=1,aplha=0.5)+
    #geom_smooth(aes(y=count))+
    scale_y_continuous(trans="log1p",breaks = c(1,2,5,10,20,50,100,200,500,1000,2000,5000),limits=c(0,NA))+
    scale_linetype(name="Source")+
    scale_colour_discrete(name="S gene status")+
    #standardPrintOutput::narrower()+
    standardPrintOutput::smallLegend(textSize = 8,spaceLegend = 1)+
    ylab("Number of cases")+
    scale_x_date(limits = c(as.Date("2021-01-21"),NA), date_breaks = "2 weeks", date_labels = "%d-%m")


  combinedPlot = ((countsBySgene+theme(legend.justification='left',legend.spacing = unit(0,"mm"),axis.text.x = element_text(angle=60))+standardPrintOutput::narrower()+standardPrintOutput::smallLegend(spaceLegend = 1)) |
    ( (countSposSeqByType+guides(colour=guide_none(),fill=guide_none())+standardPrintOutput::narrower()+standardPrintOutput::hideX()) /
    (pSposPlot+standardPrintOutput::narrower()+theme(axis.text.x = element_text(angle=60)))))+
    patchwork::plot_annotation(tag_levels = "A")
  
  combinedPlot %>% standardPrintOutput::saveHalfPageFigure(filename=paste0("~/Dropbox/covid19/sa-variant/caseCountsAndSequenceCountsBySgene-",Sys.Date()))
```


```{r}
# TODO: NHSER breakdown of panel C
# Same plots as above stratified by screening
# 
# pSposIsVariant = allVoc %>% 
#   inner_join(sgll %>% select(FINALID, sgtf_under30CT), by=c("finalid"="FINALID")) %>%
#   filter(sgtf_under30CT == 0) %>%
#   mutate(
#     detected = ifelse(asymptomatic_indicator == "Y","screening","symptoms")
#   ) %>%
#   group_by(detected, type, date=specimen_date) %>%
#   summarise(count= n()) %>% 
#   ungroup() %>%
#   tidyr::complete(detected, type,date,fill=list(count=0)) %>%
#   group_by(detected, type) %>% 
#   arrange(date) %>%
#   mutate(Roll.count = stats::filter(count,filter=rep(1,14),sides=1)) %>%
#   filter(!is.na(Roll.count)) %>% 
#   group_by(detected, date) %>% 
#   mutate(binom::binom.confint(Roll.count, sum(Roll.count), conf.level = 0.95, methods = "wilson")) %>% 
#   ungroup()
# 
# 
# 
# pSpos351Plot = ggplot(pSposIsVariant %>% filter(date > "2021-01-21" & type %in% c("unclassified","B.1.351","B.1.617.1","B.1.617.2")), aes(x=date, y=mean, ymin = lower,ymax=upper,colour=type, fill=type))+geom_ribbon(alpha=0.2,colour=NA)+geom_line()+standardPrintOutput::smallLegend()+ylab("est proportion of sequenced S+")+coord_cartesian(ylim=c(0,1))+facet_wrap(vars(detected),ncol=1)+standardPrintOutput::narrower()+theme(legend.title = element_blank())
# 
# seqCounts = sgll %>%
#   inner_join(ll2 %>% select(FINALID,asymptomatic_indicator),by="FINALID") %>% 
#   #filter(sgtf_under30ct == 0) %>%
#   group_by(sgtf_under30CT,asymptomatic_indicator,date=specimen_date) %>%
#   summarise(count= n()) %>% arrange(date) %>%
#   mutate(
#     Roll.count = stats::filter(count,filter=rep(1,14)/14),
#     context = ifelse(asymptomatic_indicator == "Y","screening","symptoms"),
#     sGene = case_when(
#       is.na(sgtf_under30CT) ~ "equivocal",
#       sgtf_under30CT == 1 ~ "negative",
#       sgtf_under30CT == 0 ~ "positive"
#     )
#   )
# 
# countsBySgeneAndDetection = ggplot(seqCounts %>% filter(date > "2020-11-01" & sGene != "equivocal"),aes(x=date,y=count, colour=context, linetype = sGene))+
#   geom_line(aes(y=Roll.count))+scale_y_continuous(trans="log1p",breaks = c(1,2,5,10,20,50,100,200,500,1000,2000,5000),limits=c(0,NA))+standardPrintOutput::narrower()+standardPrintOutput::smallLegend(spaceLegend = 0.6)
# 
# (countsBySgeneAndDetection+pSpos351Plot+patchwork::plot_layout(ncol=3)) %>% standardPrintOutput::saveHalfPageFigure(filename="~/Dropbox/covid19/sa-variant/countsBySgeneAndPVariantVsDetection")

```


## Age distributions ----

```{r}
tmp2 = allEpisodes %>% 
  left_join(llDemog, by="FINALID") %>%
  left_join(aocDf, by=c("LTLA_code"="code")) %>% 
  select(FINALID,eraIndex,date=earliest_specimen_date,sGene,area,LTLA_code,LTLA_name,age,sex,ethnicity_final,imd_decile,residential_category) %>%
  mutate(area = ifelse(is.na(area),"other",as.character(area))) %>%
  filter(date>="2020-10-01")
  
tmp2 %>% readr::write_csv(paste0("~/Dropbox/covid19/sa-variant/raw-demographics-line-list-by-area-of-concern-",Sys.Date(),".csv"))
```  


#tmp3 %>% readr::write_csv(paste0("~/Dropbox/covid19/sa-variant/age-distributions-all-areas-",Sys.Date(),".csv"))
#aocDf %>% readr::write_csv(paste0("~/Dropbox/covid19/sa-variant/area-definitions-",Sys.Date(),".csv"))


```{r}
# datePoint = (tmp2$date %>% max())-4
window = 14

dates = as.Date(c("2021-03-23","2021-04-05", "2021-04-23"))

tmp3 = tmp2 %>%
    mutate(ageCat = cut(age,c(0,10,20,30,40,50,60,70,80,Inf))) %>%
    filter(sGene %in% c("positive","negative") & !is.na(ageCat)) %>%
    mutate(dateWindow = case_when(
      date>dates[1] & date<=dates[1]+window ~ dates[1],
      date>dates[2] & date<=dates[2]+window ~ dates[2],
      date>dates[3] & date<=dates[3]+window ~ dates[3],
      TRUE ~ as.Date(NA)
    )) %>%
    filter(!is.na(dateWindow)) %>%
    group_by(dateWindow,area,sGene,ageCat) %>% summarise(count=n()) %>%
    group_by(dateWindow,area,sGene) %>% mutate(binom::binom.confint(count,sum(count),methods = "wilson"))

tmp3 %>% readr::write_csv(paste0("~/Dropbox/covid19/sa-variant/demographics-histogram-by-area-of-concern-",Sys.Date(),".csv"))

agebars = function(regions) {
  
  dodge <- position_dodge(width=0.8)
  ggplot(tmp3 %>% filter(area %in% regions), aes(x=ageCat,fill=sGene, y=mean,ymin=lower,ymax=upper))+ 
    geom_bar(stat="identity",position=dodge,width=0.8, size=0.5)+facet_grid(cols = vars(dateWindow),rows = vars(area),scales = "free_y")+
    geom_errorbar(position = dodge, width = 0.2,size=0.5)+
    #standardPrintOutput::narrower()+
    ylab("density")+theme(axis.text.x = element_text(angle=60))+xlab("age range")
  
}



#p1=agebars(c("NW Cluster","Bedford etc","Sefton & Lpl"),maxDate-21,28)+standardPrintOutput::hideX() 
# agebars(clusters2$`NW Cluster`,maxDate) %>% standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/AgeDistributions-NW Cluster-by-time")

lapply(names(clusters2), function(name) {
  agebars(clusters2[[name]]) %>% standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/AgeDistributions-",name,"-",Sys.Date()))
})

#agebars(clusters2$Midlands) %>% standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/AgeDistributions-Midlands"))

# p3=agebars(c("NW Cluster","Bedford etc","Sefton & Lpl"),maxDate-21,14)+theme(axis.text.x = element_text(angle=60))+standardPrintOutput::hideY() +xlab("age range")
# p4=agebars(c("NW Cluster","Bedford etc","Sefton & Lpl"),maxDate-14,14)+theme(axis.text.x = element_text(angle=60))+standardPrintOutput::hideY() +xlab("age range")

# (p2+p3+p4+patchwork::plot_layout(nrow=1,guides = "collect")) %>% standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/AgeDistributions-CommunityIndia2-by-time")

```


# OLD CODE ----


```{r}
# plotWithLabels = function(ltlaCounts, xlim = c(limits[2],limits[4]), ylim=c(limits[1],limits[3]), limits = c(NA,NA,NA,NA), labelsPerPlot = 4, nrow=2) {
#   
#   ltlaCounts = ltlaCounts %>% filter(category %in% c("unsequenced S+","B.1.351","B.1.617.2"))
#   ladMap = dpc$geog$getMap("LAD19")
#   ltlas = ladMap %>% as_tibble() %>% select(code,name) %>% distinct() %>%
#     mutate(label = code %>% stringr::str_replace("E0(.)0+(.+)","\\1.\\2"))
#   categories = ltlaCounts %>% select(category) %>% distinct() %>% arrange(category)
# 
#   ltlaCounts2 = ltlas %>% select(-name) %>% inner_join(categories, by=character()) %>%
#     left_join(ltlaCounts %>% select(-name), by=c("code","category")) %>% 
#     mutate(count = ifelse(is.na(count),0,count))
#   
#   mapVoc = ltlaCounts2 %>% left_join(ladMap, by=c("code"), suffix=c("",".map")) %>% 
#     filter(code %>% stringr::str_starts("E")) %>% sf::st_as_sf()
#   
#   labels = mapVoc %>% 
#     sf::st_centroid() %>%
#     mutate(x=sf::st_coordinates(.)[,"X"], y=sf::st_coordinates(.)[,"Y"]) %>% 
#     filter(x > ifna(xlim[1],-Inf) & x < ifna(xlim[2],Inf) ) %>%
#     filter(y > ifna(ylim[1],-Inf) & y < ifna(ylim[2],Inf) ) %>% 
#     as_tibble() %>%
#     group_by(category) %>% 
#     arrange(desc(count)) %>% 
#     filter(row_number() <= labelsPerPlot) %>%
#     ungroup()
#   
#   p1 = lapply(categories$category, function(cat) {
#     breaks = mapVoc %>% filter(category==cat & count>0) %>% pull(count) %>% quantile(c(0.5,0.75,0.9,0.95,0.975))
#     breaks = sort(unique(round(c(1,unname(breaks[breaks>1])))))
#     plotlabels = c("0",ifelse(c(0,breaks)+1 == c(breaks,Inf), as.character(breaks) , paste0( as.character(c(0,breaks)+1),c(rep("-",length(breaks)),"+"),c(as.character(breaks),""))))
#     tmp = mapVoc %>% filter(category==cat) %>% mutate(count2=cut(count,breaks = c(-Inf,0,breaks,Inf),labels = plotlabels,include.lowest = FALSE))
#     #browser()
#     ggplot()+
#     #geom_sf(data=dpc$geog$getMap("LAD19") ,size=0.05)+
#       geom_sf(data=tmp,mapping=aes(fill=count2),size=0.1,colour="white")+
#       scale_fill_viridis_d(option = "inferno", name="")+
#       standardPrintOutput::defaultMapLayout()+
#       standardPrintOutput::narrower()+
#       standardPrintOutput::smallLegend(textSize = 6)+ #spaceLegend = 0.8)+
#       facet_wrap(vars(category))+
#       coord_sf(xlim = xlim,ylim= ylim)+
#       ggrepel::geom_label_repel(data = labels %>% filter(category==cat), aes(x=x,y=y,label=label),min.segment.length = 0,segment.colour = "cyan",fill="#80808080",colour="cyan",size=1.5)
#   })
#   
#   
#   #print(labs %>% mutate(l = paste0(label,": ",name)) %>% pull(l) %>% paste0(collapse = "; "))
#   p4 = standardPrintOutput::simpleFigureTable(labels %>% as_tibble() %>% select(label,name) %>% distinct() %>% arrange(label),pts = 5,unwrapped = TRUE)
#   #browser()
#   return(patchwork::wrap_plots(c(p1,list(p4)),nrow=nrow)+patchwork::plot_annotation(title = paste0("cases between ",earliestDate," and ",lastDate))) #, widths = c(3,3,3,2)))#+patchwork::plot_layout(ncol = 2))
# }
```


