---
title: "B.1.251 sequences and S-gene positive pillar 2 cases"
author: "Rob Challen"
date: '`r format(Sys.Date(), "%d-%m-%Y")`'
output: 
  pdf_document :
    fig_caption: yes
header-includes:
  \usepackage{float}
  \floatplacement{figure}{H}
knit: (function(inputFile, encoding,...) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "~/Dropbox/covid19/sa-variant/")})
fig_width: 7
fig_height: 5
out.width: "100%"
---


```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo = FALSE,
  warning = FALSE,
  message = FALSE
)

devtools::load_all("~/Git/standard-print-output/")
devtools::load_all("~/Git/uk-covid-datatools/")

library(tidyverse)
library(patchwork)
library(rgdal)
library(ggplot2)
library(ggspatial)
library(rgeos)
library(maptools)
library(patchwork)
library(sp)
library(sf)

ggplot2::theme_set(standardPrintOutput::defaultFigureLayout())
standardPrintOutput::setDefaults()
ukcovidtools::setup()
ukcovidtools::reload()
```

```{r eval=FALSE}
devtools::load_all("~/Git/uk-covid-datatools/")
ukcovidtools::reload()

setwd("~/Git/uk-covid-datatools/")
if (!exists("combinedSpositives")) source("./vignettes/s-pos-voc/s-pos-data.R")
```

# Data set for S-gene growth rates

* How do we construct the case count of S-positive and S-negative cases by area of concern & other & England?
* problem in that sgene line list is test level and contains >1 test per person
* scale of this problem
* sgll %>% filter(specimen_date>"2021-03-01") %>% group_by(FINALID,sgtf_under30CT) %>% summarise(count=n()) %>% group_by(sgtf_under30CT,count) %>% summarise(freq = n())

1) line list left join s-gene line list by final id and specimen_date. filter on S-gene status defined (pos,neg,equivocal).
* this is most naive approach and will give you the subset of Pillar 2 first infections that have a s-gene status.
* will lose people who have a first infection that does not have a taqpath result
* can make this a bit flexible in terms of date matching (s-gene date within 4 days line list date...) 
* either way still need to deduplicate by finalid, in case there was >1 test for that finalid on that day.  in which case must also de-duplicate. 
* cannot tell you about re-infections.
* Will lose people that have long infection episode with a later test being done at TaqPath
* can tell you about Pillar 1 & 2 tests where taqpath results not available.
* FIRST_INFECTION

2) s-gene line list left-join demographics from line list, - order by date and get first occurrence for each finalid. Filter by S-positive (sgtf_under30CT == 0).
* probably gives you similar to 1) but removes need for S-gene to be first ever test for an individual. 
* Instead gives first ever taqpath test for patient, which might not be the first test in an infection episode.
* Will ensure uniqueness - only one test per patient.
* Nothing for re-infections
* FIRST_TAQPATH

3) s-gene line list left-join demographics from line list, - order by reverse date and get last occurrence for each finalid. Filter by S-positive.
* Last ever taqpath test for patient.
* Dates for patients with >1 test shifted backwards.
* if test represents 2 infection episodes will give you most recent episode
* LAST_TAQPATH

4) combined ll & sgll line-list with infection episodes defined as per s-pos-data allEpisodes
* Date as per earliest_specimen_date
* Episodes are unique finalid+eraIndex but not finalid alone
* Potential re-infection detected.
* May potentially shift people much earlier than other methods if they have had a few S-gene unknown P1 & P2 tests followed by TaqPath test.
* INFECTION_EPISODES



# 1) FIRST_INFECTION

```{r}

firstInfection = ll %>% left_join(sgll, by="FINALID", suffix=c("",".sgene")) %>% 
  filter(is.na(specimen_date.sgene) | (specimen_date <= specimen_date.sgene & specimen_date+4 > specimen_date.sgene)) %>% 
  arrange(FINALID) %>% filter(FINALID != lag(FINALID)) %>% mutate(
    sGene = case_when(
      sgtf_under30CT == 1 ~ "negative",
      sgtf_under30CT == 0 ~ "positive",
      is.na(sgtf) ~ "unknown",
      is.na(sgtf_under30CT) ~ "equivocal",
      TRUE ~ NA_character_
    )
  ) %>% select(all_of(colnames(ll)),sGene)

if(any(duplicated(firstInfection$FINALID))) stop("duplicates detected")

firstInfection %>% with(table(asymptomatic_indicator,sGene))

```

# 2) FIRST_TAQPATH

```{r}

firstTaqpath = sgll %>% left_join(llDemog, by="FINALID") %>%
  left_join(llTest, by=c("FINALID","specimen_date")) %>% 
  filter(!is.na(specimen_date)) %>% 
  arrange(FINALID,specimen_date) %>% 
  filter(FINALID != lag(FINALID)) %>% 
  mutate(
    pillar="Pillar 2",
    sGene = case_when(
      sgtf_under30CT == 1 ~ "negative",
      sgtf_under30CT == 0 ~ "positive",
      is.na(sgtf) ~ "unknown",
      is.na(sgtf_under30CT) ~ "equivocal",
      TRUE ~ NA_character_
    ),
    asymptomatic_indicator = ifelse(is.na(asymptomatic_indicator),"U",asymptomatic_indicator)
  )

if(any(duplicated(firstTaqpath$FINALID))) stop("duplicates detected")

firstTaqpath %>% with(table(asymptomatic_indicator,sGene))

```

# 3) LAST_TAQPATH

```{r}

lastTaqpath = sgll %>% left_join(llDemog, by="FINALID") %>% 
  left_join(llTest, by=c("FINALID","specimen_date")) %>% 
  filter(!is.na(specimen_date)) %>% 
  arrange(FINALID,desc(specimen_date)) %>% 
  filter(FINALID != lag(FINALID)) %>% 
  mutate(
    pillar="Pillar 2",
    sGene = case_when(
      sgtf_under30CT == 1 ~ "negative",
      sgtf_under30CT == 0 ~ "positive",
      is.na(sgtf) ~ "unknown",
      is.na(sgtf_under30CT) ~ "equivocal",
      TRUE ~ NA_character_
    ),
    asymptomatic_indicator = ifelse(is.na(asymptomatic_indicator),"U",asymptomatic_indicator)
  )

if(any(duplicated(lastTaqpath$FINALID))) stop("duplicates detected")

lastTaqpath %>% with(table(asymptomatic_indicator,sGene))

```
# 4) INFECTION_EPISODES

```{r}
infectionEpisodes = allEpisodes %>% inner_join(llDemog,by="FINALID")%>% rename(specimen_date=earliest_specimen_date)

infectionEpisodes %>% with(table(asymptomatic_indicator,sGene))

```
```{r}




buildTimeseries = function(filterExpr = TRUE) {
  
  filterExpr = enexpr(filterExpr)
  
  lastTrustworthyDate = min(c(
    max(sgll$specimen_date,na.rm = TRUE),
    max(ll$specimen_date,na.rm = TRUE),
    max(allEpisodes$earliest_specimen_date)
  ))-4
  
  tmp = bind_rows(
    firstInfection %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "LAD", subgroup = sGene, filterExpr = !!filterExpr) %>% mutate(source="first infection"),
    firstTaqpath %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "LAD", subgroup = sGene, filterExpr = !!filterExpr) %>% mutate(source="first taqpath"),
    lastTaqpath %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "LAD", subgroup = sGene, filterExpr = !!filterExpr) %>% mutate(source="last taqpath"),
    infectionEpisodes %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "LAD", subgroup = sGene, filterExpr = !!filterExpr) %>% mutate(source="infection episodes")
  )
  
  tmp = tmp %>% dpc$demog$findDemographics()
  
  tmp2 = tmp %>% 
    left_join(aocDf %>% select(area,code),by = "code") %>% 
    mutate(area = forcats::fct_explicit_na(area,na_level = "other")) %>% 
    select(-code,-name) %>% 
    mutate(code=area,name=area) %>% 
    covidStandardDateGrouping() %>% 
    summarise(
      value = sum(value),
      population = sum(population,na.rm = TRUE),
      Implicit = all(Implicit)
    ) 
  
  
  tmp3 = bind_rows(
    firstInfection %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "CTRY", subgroup = sGene, filterExpr = !!filterExpr) %>% mutate(source="first infection"),
    firstTaqpath %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "CTRY", subgroup = sGene, filterExpr = !!filterExpr) %>% mutate(source="first taqpath"),
    lastTaqpath %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "CTRY", subgroup = sGene, filterExpr = !!filterExpr) %>% mutate(source="last taqpath"),
    infectionEpisodes %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "CTRY", subgroup = sGene, filterExpr = !!filterExpr) %>% mutate(source="infection episodes")
  )
  
  tmp3 = tmp3 %>% dpc$demog$findDemographics()
  
  tmp4 =   bind_rows(tmp2,tmp3) %>% filter(date <= lastTrustworthyDate)
  return(tmp4)
}

allTimeseries = buildTimeseries()
symptomaticTimeseries = buildTimeseries(asymptomatic_indicator != "Y")

allTimeseries %>% readr::write_csv(paste0("~/Dropbox/covid19/sa-variant/all-case-counts-by-area-of-concern-and-method-",Sys.Date(),".csv"))
symptomaticTimeseries %>% readr::write_csv(paste0("~/Dropbox/covid19/sa-variant/symptomatic-case-counts-by-area-of-concern-and-method-",Sys.Date(),".csv"))

```



# Integrate data from Laura and Chris

```{r}


laura1 = readr::read_csv("~/Dropbox/covid19/sa-variant/Inputs/R18pIV_31052021_GrowthRate_all_prior2_2_119.csv")
laura2 = readr::read_csv("~/Dropbox/covid19/sa-variant/Inputs/R18pIV_31052021_GrowthRate_symptomatic_prior2_2_119.csv")
chris = readr::read_csv("~/Dropbox/covid19/sa-variant/Inputs/combined_results_UoM-2021-06-01.csv")
david = readr::read_csv("~/Dropbox/covid19/sa-variant/Inputs/output_complete-2021-06-02.csv")

laura3 = bind_rows(
  laura2 %>% mutate(dataset = "Symptomatic") %>% select(-code) %>% rename(subgroup2 = subgroup,subgroup=source) %>% rename(source=subgroup2),
  laura1 %>% mutate(dataset = "All") %>% select(-code) %>% rename(subgroup2 = subgroup,subgroup=source) %>% rename(source=subgroup2)
) %>% 
mutate(
  Growth.median = median, 
  Growth.lower = q0.025,
  Growth.upper = q0.975
) %>% select(-starts_with("q"))
  
chris2 = chris %>% 
  pivot_longer(cols = c(-Region,-Type,-Source,-Status,-Parameter),names_to="date",values_to="value") %>%
  pivot_wider(names_from = "Parameter",values_from="value") %>% rename(
    name = Region,
    subgroup = Type,
    source = Source,
    dataset = Status,
    Growth = `mean growth rate`,
    Growth.SE = `standard deviation`
  ) %>% mutate(
    Growth.median = Growth,
    Growth.lower = Growth-1.96*Growth.SE,
    Growth.upper = Growth+1.96*Growth.SE,
    date = as.Date(date,"%d/%m/%Y"),
    source = stringr::str_to_lower(source),
    subgroup = stringr::str_remove(subgroup,"S ")
  )

david2 = david %>% select(
  date,name, subgroup, source, 
  Growth.median = MAP,
  Growth.lower = lower,
  Growth.upper = upper
) %>% mutate(
  dataset = "Symptomatic"
)

```

Modelled growth rate

```{r}

lsoaCentroid = dpc$geog$getMap("LSOA11")
lsoaCentroid = lsoaCentroid %>% sf::st_centroid()

mapping = dpc$geog$getContainedIn(inputSf = lsoaCentroid,outputShape = aocMap,outputIdVar = areaName)
mapping %>% readr::write_csv("~/Dropbox/covid19/sa-variant/Inputs/LSOAtoAOCMapping.csv")

mapping = readr::read_csv("~/Dropbox/covid19/sa-variant/LSOAtoAOCMapping.csv")
modelled = readr::read_csv("~/Dropbox/covid19/sa-variant/Inputs/ByLSOAVariantRates-2021-06-02.csv")

modelledAoc = bind_rows(
  modelled %>% inner_join(mapping, by=c("LSOA_code"="from")) %>% group_by(area = to, week, type) %>% summarise(variantGR = sum(VariantRate)),
  modelled %>% filter(LSOA_code %>% stringr::str_starts("E")) %>% group_by(week, type) %>% summarise(variantGR = sum(VariantRate,na.rm = TRUE)) %>% mutate(area="England")
)

estGR = modelledAoc %>% group_by(area,type) %>% arrange(week) %>%
  mutate(
    gr = (log(variantGR)-log(lag(variantGR)))/7,
    date = week+as.numeric(week-lag(week))/2
  )

#ggplot(estGR,aes(x=date,y=gr,colour=type))+ geom_point()+facet_wrap(vars(area))
doModelBar = function(clustName = paste(names,collapse = "-"), names = clusters[[clustName]]) {
  ggplot(data=estGR %>% filter(area %in% names) %>% rename(name=area),mapping=aes(x=date,y=gr,fill=type))+
    geom_bar(stat="identity",position = "dodge", width=5, colour="black",size=0.1)+
    scale_y_continuous(limits = c(-0.25,0.25))+
    scale_fill_brewer(palette="Set2",name=NULL)+
    scale_x_date(breaks=seq(as.Date("2021-02-04"),Sys.Date(),by = 7),date_labels="%d-%m")+
    ylab("growth rate, r")
}

p1 = doModelBar(names = "England")+
  facet_wrap(vars(name),nrow=1)+
  theme(axis.text.x = element_text(angle=0,hjust=0.5))+
  guides(linetype=guide_none(),colour=guide_none(),fill=guide_none())+
  theme(axis.title.x = element_blank())+
  theme(plot.margin = margin(0, 0, 0, 0, "mm"))

p2 = doModelBar(clustName = "NHS Regions")+
  facet_wrap(vars(name),nrow=1)+
  guides(linetype=guide_none(),colour=guide_none(),fill=guide_none())+
  standardPrintOutput::hideX()+
  theme(plot.margin = margin(0, 0, 0, 0, "mm")) #,axis.ticks.x = element_blank())

p3 = doModelBar(names = c("Bolton","M65 Corridor","Manchester"))+
  facet_wrap(vars(name),nrow=1)+
  guides(linetype=guide_none(),colour=guide_none(),fill=guide_none())+
  standardPrintOutput::hideX()+
  theme(plot.margin = margin(0, 0, 0, 0, "mm")) #,axis.ticks.x = element_blank())

p4 = doModelBar(names = c("Leicester","Nottingham","Birmingham"))+
  facet_wrap(vars(name),nrow=1)+
  standardPrintOutput::smallLegend(spaceLegend = 1)+theme(
    legend.position = "bottom",
    legend.direction = "horizontal"
  )

p5 = (p1/p2/p3/p4 + patchwork::plot_annotation(tag_levels="A"))+patchwork::plot_layout(heights = c(2,2,1.5,1.5))
standardPrintOutput::saveTwoThirdPageFigure(p5,paste0("~/Dropbox/covid19/sa-variant/ModelledGrowth-",Sys.Date()))

```

# Calculate growth rate ----

```{r}

devtools::load_all("~/Git/uk-covid-datatools/")
ukcovidtools::reload()

allTimeseriesGR = allTimeseries %>% 
  tsp$estimateGrowthRate(window = 56,leftSided=FALSE) %>%
  tsp$estimateGrowthRate2(window = 14) #smoothingWindow = 14,leftSided=FALSE,nocache=TRUE)
  

symptomaticTimeseriesGR = symptomaticTimeseries %>% 
  tsp$estimateGrowthRate(window = 56,leftSided=FALSE) %>%
  tsp$estimateGrowthRate2(window = 28, nocache=TRUE)
  #tsp$logIncidenceStats(smoothingWindow = 14,leftSided=FALSE,nocache=TRUE) 

pois = bind_rows(
  allTimeseriesGR %>% mutate(dataset = "All"),
  symptomaticTimeseriesGR  %>% mutate(dataset = "Symptomatic")
) %>% select(
  date,name, subgroup, source, dataset, Growth = Growth.poisson, Growth.SE = Growth.SE.poisson
) %>% mutate(
  Growth.median = Growth,
  Growth.lower = Growth-1.96*Growth.SE,
  Growth.upper = Growth+1.96*Growth.SE
) 

logIncid = bind_rows(
  allTimeseriesGR %>% mutate(dataset = "All"),
  symptomaticTimeseriesGR  %>% mutate(dataset = "Symptomatic")
) %>% select(
  date,name, subgroup, source, dataset, Growth.Quantile.0.5.value, Growth.Quantile.0.025.value, Growth.Quantile.0.975.value
) %>% rename(
  Growth.median = Growth.Quantile.0.5.value,
  Growth.lower = Growth.Quantile.0.025.value,
  Growth.upper = Growth.Quantile.0.975.value
)

```


```{r}
devtools::load_all("~/Git/uk-covid-datatools/")
ukcovidtools::reload()

symptomaticTimeseriesGR = symptomaticTimeseries %>% 
  tsp$estimateGrowthRate(window = 56,leftSided=FALSE) %>%
  tsp$estimateGrowthRate2(window = 28, nocache=TRUE)

logIncid = bind_rows(
  allTimeseriesGR %>% mutate(dataset = "All"),
  symptomaticTimeseriesGR  %>% mutate(dataset = "Symptomatic")
) %>% select(
  date,name, subgroup, source, dataset, Growth.Quantile.0.5.value, Growth.Quantile.0.025.value, Growth.Quantile.0.975.value
) %>% rename(
  Growth.median = Growth.Quantile.0.5.value,
  Growth.lower = Growth.Quantile.0.025.value,
  Growth.upper = Growth.Quantile.0.975.value
)

allEstimates = bind_rows(
  laura3 %>% mutate(method = "GP"),
  chris2 %>% mutate(method = "GAM"),
  pois %>% mutate(method = "Pois (GLM)"),
  david2 %>% mutate(method = "Pois (Bayes)"),
  logIncid %>% mutate(method = "Pois (Loess)")
)

p3 = doGRPlot(allEstimates,names = c("Bolton","M65 Corridor","Manchester"),mergeModels = TRUE)+
  #doModelPlot(names = c("Bolton","M65 Corridor","Manchester"))+
  facet_wrap(vars(name),nrow=1)+
  guides(linetype=guide_none(),colour=guide_none(),fill=guide_none())+
  standardPrintOutput::hideX()+
  theme(plot.margin = margin(0, 0, 0, 0, "mm")) #,axis.ticks.x = element_blank())

p3
```
<!-- Debug M65 Corridor -->

<!-- ```{r} -->

<!-- tmp = symptomaticTimeseries %>% filter(source == "first infection" & subgroup == "positive" & code == "M65 Corridor") %>% arrange(date) -->


<!-- for (i  in 1:nrow(tmp)) { -->
<!--   tmp2 = tmp %>% filter(row_number() > i-28 & row_number()<i+28) -->
<!--   tmp2 = tmp2 %>% rename(I = value) %>% mutate(time = row_number()) -->
<!--   tmp3 = tryCatch({ -->
<!--       model <- suppressWarnings(glm(I ~ time, family = quasipoisson(), data = tmp2)) -->
<!--       model_sum <- summary(model) -->
<!--       print(paste0("gr: ",model_sum$coefficients[2, 1],": ",paste0(tmp2$I, collapse=", "))) -->
<!--   }, error = browser) -->
<!-- } -->


<!-- I_vec = c(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0) -->
<!-- tmp2 = tibble(time = 1:length(I_vec), I = I_vec) -->
<!-- model <- suppressWarnings(glm(I ~ time, family = quasipoisson(), data = tmp2)) -->
<!-- summary(model) -->

<!-- ``` -->


```{r}

allAreas = unname(c(unlist(clusters)))

allEstimates = bind_rows(
  laura3 %>% mutate(method = "GP"),
  chris2 %>% mutate(method = "GAM"),
  pois %>% mutate(method = "Pois (GLM)"),
  david2 %>% mutate(method = "Pois (Bayes)"),
  logIncid %>% mutate(method = "Pois (Loess)")
)

growthRateTest = list(
  input = symptomaticTimeseries %>% filter(source == "infection episodes") %>% ungroup() %>% as_tibble() %>% select(-type,-statistic,-ageCat,-gender,-codeType,-code,-Implicit, -name.original, -note), 
  output = allEstimates %>% filter(source == "infection episodes" & dataset == "Symptomatic") %>% select(-dataset,-median)
)

usethis::use_data(growthRateTest,overwrite = TRUE)

doGRPlot = function(
  allEstimates,
  sources="infection episodes",
  minDate="2021-01-01",
  subgroups = c("positive","negative"),
  datasets="Symptomatic",
  methods = unique(allEstimates$method),
  clustName = paste(names,collapse = "-"),
  names = clusters[[clustName]],
  mergeModels = FALSE
  ) {
  
    filteredTS = allEstimates %>% 
      filter(source %in% sources & date>minDate) %>% 
      filter(subgroup %in% subgroups) %>%
      filter(dataset %in% datasets) %>%
      filter(method %in% methods) %>% 
      filter(name %in% names) %>% 
      mutate(name = name %>% ordered(names))
    
    modelEndDate = filteredTS %>% filter(!is.na(Growth)) %>% group_by(source,subgroup,dataset,name,method) %>% summarise(end = max(date,na.rm = TRUE)) %>% pull(end) %>% max(na.rm = TRUE)
    #browser()
    filteredTS = filteredTS %>% filter(date<=modelEndDate)
    
    # if (mergeModels) {
    #   
    #   filteredTS = filteredTS %>% 
    #     group_by(source,subgroup,dataset,name, date) %>%
    #     group_modify(function(d,g,...) {
    #       mus = d$Growth
    #       sigmas = d$Growth.SE
    #       boots = length(mus)
    #       cdf = function(x) LaplacesDemon::pnormm(x, p=rep(1/boots,boots), mu=mus, sigma=sigmas)
    #       # solve CDF for quantiles.
    #       quantGenFn = function(q) return(function(x) cdf(x)-q)
    #       bounds = c(-1,1) #log(c(0.5,5))
    #       mean.estimate = mean(mus)
    #       estimate.lower = tryCatch(uniroot(quantGenFn(0.025), interval=bounds)$root,error = function(e) -Inf)
    #       estimate.median = tryCatch(uniroot(quantGenFn(0.5), interval=bounds)$root,error = function(e) NaN)
    #       estimate.upper = tryCatch(uniroot(quantGenFn(0.975), interval=bounds)$root,error = function(e) Inf)
    #       return(tibble(
    #         Growth.median = estimate.median,
    #         Growth.lower = estimate.lower,
    #         Growth.upper = estimate.upper,
    #         method = "ensemble"
    #       ))
    #     }) %>% ungroup()
    # }
    
    modelComp = ggplot(filteredTS, mapping=aes(x=date))+
      geom_ribbon(mapping=aes(
        ymin=Growth.lower,
        ymax=Growth.upper,
        group=paste0(subgroup,method,dataset)),
        colour=NA, 
        fill="grey60", 
        alpha=0.2)+
      geom_line(mapping=aes(
        y=Growth.median,
        color = subgroup,
        linetype=method)
      )+ 
      scale_y_continuous(
        sec.axis = dup_axis( 
          breaks = log(2)/c(3,7,14,Inf,-14,-7,-3), 
          labels = c("3","7","14","\u221E","-14","-7","-3"), 
          name="doubling time")
      )+
      scale_x_date(breaks=seq(as.Date("2021-02-04"),Sys.Date(),by = 14),date_labels="%d-%m")+  #date_breaks = "2 week",date_labels="%d-%m")+
      scale_color_discrete(name="S-gene")+
      coord_cartesian(ylim=c(-0.25,0.25), xlim=c(as.Date("2021-02-01"),NA))+
      ylab("growth rate, r")+
      theme(axis.text.x = element_text(angle=60), axis.text.y.right = element_text(size=7))+
      geom_hline(yintercept = 0, colour="grey50")+
      xlab("specimen date")
    
    #if (!mergeModels) 
      modelComp = modelComp + scale_linetype_manual(values=c("Pois (GLM)"="solid","GAM"="twodash", "GP"="dotted","Pois (Bayes)"="dotdash","Pois (Loess)"="longdash"))
      
    modelComp
}
```


```{r}
# lapply(names(clusters),function(clustName) {
#   allEstimates %>% 
#     doGRPlot(clustName = clustName) %>% 
#     standardPrintOutput::saveThirdPageFigure(paste0("~/Dropbox/covid19/sa-variant/sGeneAOCHighGrowth-",clustName))})
```

# FIGURE 2

```{r}

doModelPlot = function(clustName = paste(names,collapse = "-"), names = clusters[[clustName]]) {
  list( #ggnewscale::new_scale_colour(),
    geom_point(data=estGR %>% filter(area %in% names & type %in% c("B.1.1.7","B.1.617.2") & date<=max(allEstimates$date)) %>% rename(name=area),mapping=aes(x=date,y=gr,fill=type),shape=21,stroke=0),
    shades::lightness(scale_fill_discrete(), shades::scalefac(0.70))
  )
}

p1 = doGRPlot(allEstimates,names = "England")+
  #doModelPlot(names = "England")+
  facet_wrap(vars(name),nrow=1)+
  theme(axis.text.x = element_text(angle=0,hjust=0.5))+
  guides(linetype=guide_none(),colour=guide_none(),fill=guide_none())+
  theme(axis.title.x = element_blank())+
  theme(plot.margin = margin(0, 0, 0, 0, "mm"))

p2 = doGRPlot(allEstimates,clustName = "NHS Regions",mergeModels = TRUE)+
  #doModelPlot(clustName = "NHS Regions")+
  facet_wrap(vars(name),nrow=1)+
  guides(linetype=guide_none(),colour=guide_none(),fill=guide_none())+
  standardPrintOutput::hideX()+
  theme(plot.margin = margin(0, 0, 0, 0, "mm")) #,axis.ticks.x = element_blank())

p3 = doGRPlot(allEstimates,names = c("Bolton","M65 Corridor","Manchester"),mergeModels = TRUE)+
  #doModelPlot(names = c("Bolton","M65 Corridor","Manchester"))+
  facet_wrap(vars(name),nrow=1)+
  guides(linetype=guide_none(),colour=guide_none(),fill=guide_none())+
  standardPrintOutput::hideX()+
  theme(plot.margin = margin(0, 0, 0, 0, "mm")) #,axis.ticks.x = element_blank())

p4 = doGRPlot(allEstimates,names = c("Leicester","Nottingham","Birmingham"),mergeModels = TRUE)+
  #doModelPlot(names = c("Leicester","Nottingham","Birmingham"))+
  facet_wrap(vars(name),nrow=1)+
  standardPrintOutput::smallLegend(spaceLegend = 1)+theme(
    legend.position = "bottom",
    legend.direction = "horizontal"
  )

p5 = (p1/p2/p3/p4 + patchwork::plot_annotation(tag_levels="A"))+patchwork::plot_layout(heights = c(2,2,1.5,1.5))
#standardPrintOutput::saveTwoThirdPageFigure(p5,paste0("~/Dropbox/covid19/sa-variant/sGeneAOCHighGrowth-Fig2-",Sys.Date()))
standardPrintOutput::saveTwoThirdPageFigure(p5,paste0("~/Dropbox/covid19/sa-variant/sGeneAOCHighGrowth-nomodel-Fig2-",Sys.Date()))

allEstimates %>% filter(source=="infection episodes" & date>"2021-01-01" & subgroup %in% c("positive","negative") & dataset=="Symptomatic") %>% 
  readr::write_csv(paste0("~/Dropbox/covid19/sa-variant/allEstimates-Fig2-",Sys.Date(),".csv"))
```

```{r}

tmp = doGRPlot(allEstimates,
         sources = unique(allEstimates$source),
         datasets = "Symptomatic",
         names = allAreas)+
  facet_grid(name~source)+
  standardPrintOutput::smallLegend(spaceLegend = 1)+theme(
    legend.position = "bottom",
    legend.direction = "horizontal"
  )

standardPrintOutput::saveFullPageFigure(tmp,paste0("~/Dropbox/covid19/sa-variant/sGeneAOCSensitivity-to-sgene-method-",Sys.Date()))
```

```{r}
tmp = doGRPlot(allEstimates,
         sources = "infection episodes",
         datasets = c("All","Symptomatic"),
         names = allAreas)+
  facet_wrap(vars(name,dataset),ncol=4)+
  standardPrintOutput::smallLegend(spaceLegend = 1)+theme(
    legend.position = "bottom",
    legend.direction = "horizontal"
  )

standardPrintOutput::saveFullPageFigure(tmp,paste0("~/Dropbox/covid19/sa-variant/sGeneAOCSensitivity-to-dataset-",Sys.Date()))

```


```{r}

# doGRPlot(allEstimates,methods="Pois",names = allAreas,sources = unique(allEstimates$source))+facet_wrap(vars(name,source))

```

```{r}

# plotAllGR = function(dataGR) {
#   lapply(names(clusters),function(clustName) {
#     (
#       tsp$plotGrowthRate(
#         dataGR %>% filter(name %in% clusters[[clustName]]) %>% filter(subgroup!="equivocal"),
#         colour = subgroup,dates="2021-03-01",
#         rlim=c(-0.25,0.25),
#         growthVar = Growth.poisson,
#         growthSEVar = Growth.SE.poisson,
#       )+facet_wrap(vars(name))+standardPrintOutput::watermark(lab="provisional")+ylab("r - sliding 8 week window")
#     ) %>%   
#     standardPrintOutput::saveThirdPageFigure(paste0("~/Dropbox/covid19/sa-variant/sGeneAOCHighGrowth-",clustName))
#   })
# }
# 
# 
# 
# 
# aocSgeneCountsRt = aocSgeneCounts %>% tsp$estimateRt(window = 14)
# lapply(names(clusters),function(clustName) {
#   (
#     tsp$plotRt(
#       aocSgeneCountsRt %>% filter(name %in% clusters[[clustName]]) %>% filter(subgroup!="equivocal"),
#       colour = subgroup,
#       dates="2021-03-01",
#       rtlim=c(0,3.5)
#     )+
#     facet_wrap(vars(name))+
#     standardPrintOutput::watermark(lab="provisional")) %>%   
#   standardPrintOutput::saveThirdPageFigure(paste0("~/Dropbox/covid19/sa-variant/sGeneAOCHighRt-",clustName))
# })

# allTimeseries %>% filter(date>"2021-03-01") %>% tsp$plotIncidenceQuantiles(denominatorExpr = population/1000000,colour = source,ylim = c(0,200))+facet_grid(cols = vars(subgroup),rows=vars(name))
# allTimeseries %>% tsp$plotGrowthRate(colour = subgroup,rlim = c(-0.25,0.25),dates = "2021-03-01",linetype=source,ribbons = FALSE)+facet_wrap(vars(name))
# symptomaticTimeseries %>% tsp$plotGrowthRate(colour = subgroup,rlim = c(-0.25,0.25),dates = "2021-03-01",linetype=source,ribbons = FALSE)+facet_wrap(vars(name))
```

## R advantage

```{r}

psi = ParametricSerialIntervalProvider$new(dpc, "gamma", tibble(param = c("mean","sd"), mean=c(4.2,4.9), sd=c(NA,NA), lower=c(NA,NA), upper=c(NA,NA)))

sRT14 = symptomaticTimeseries %>% filter(name != "Unknown (England)") %>% tsp$imputeAndWeeklyAverage() %>% tsp$estimateRt(quick = TRUE,serialIntervalProvider = psi,valueVar = Imputed.value,window = 21)

tsp$plotIncidenceQuantiles(sRT14 %>% filter(source=="infection episodes"),colour = subgroup,dates = "2021-03-01",denominatorExpr = population/1000000,ylim = c(0,500))+facet_wrap(vars(name))+scale_y_continuous(trans="log1p")

tsp$plotRt(sRT14 %>% filter(source=="infection episodes"),colour = subgroup,dates = "2021-03-01",rtlim = c(0.5,2.5))+facet_wrap(vars(name))

pos = sRT14 %>% 
  filter(subgroup=="positive") %>% 
  mutate(posShape = (`Mean(R)`^2)/(`Std(R)`^2), posScale=(`Std(R)`^2)/`Mean(R)`) %>% 
  ungroup() %>%
  select(date,source,name, posShape, posScale, posMean = `Mean(R)`)

neg = sRT14 %>% 
  filter(subgroup=="negative") %>% 
  mutate(negShape = (`Mean(R)`^2)/(`Std(R)`^2), negScale=(`Std(R)`^2)/`Mean(R)`) %>% 
  ungroup() %>%
  select(date,source,name, negShape, negScale, negMean = `Mean(R)`)

rtRatio = pos %>% left_join(neg, by=c("date","source","name"))

rtRatio2 = rtRatio %>% 
  filter(!is.na(posShape) & !is.na(negShape) & !is.na(posScale) & !is.na(negScale) & negScale > 0) %>%
  mutate(
    ratio.median = gbeta::qgbetap(p=0.5,posShape,posScale/negScale,negShape,1),
    ratio.q.025 = gbeta::qgbetap(0.025,posShape,posScale/negScale,negShape,1),
    ratio.q.975 = gbeta::qgbetap(0.975,posShape,posScale/negScale,negShape,1),
    ratio.mean = posMean/negMean
  )

rtRatio3 = rtRatio %>% filter(!is.na(posShape) & !is.na(negShape) & !is.na(posScale) & !is.na(negScale) & negScale > 0) %>%
  group_by(date,source,name) %>%
  group_modify(function(d,g,...) {
    posRt = rgamma(100,shape=d$posShape,scale=d$posScale)
    negRt = rgamma(100,shape=d$negShape,scale=d$negScale)
    ratios = posRt/negRt
    tibble(
      ratio.mean = mean(ratios),
      ratio.q.025 = unname(quantile(ratios,0.025)),
      ratio.median = unname(quantile(ratios,0.5)),
      ratio.q.975 = unname(quantile(ratios,0.975))
    )
  })
  

ggplot(rtRatio3 %>% filter(source=="infection episodes" & date>"2021-03-01"),aes(x=date,y=ratio.median,ymin=ratio.q.025,ymax=ratio.q.975,group=source))+
  geom_ribbon(fill="grey",alpha=0.5)+
  #geom_line()+
  geom_line(aes(y=ratio.mean),colour = "blue")+
  geom_hline(yintercept = 1,colour="red")+
  facet_wrap(vars(name))+coord_cartesian(ylim=c(0,3))+
  ylab(latex2exp::TeX("$R_t^{S+}/R_t^{S-}$"))+
  standardPrintOutput::watermark(lab="PROVISIONAL")

#https://stats.stackexchange.com/questions/207595/distribution-of-the-ratio-of-two-gamma-random-variables
# X = gamma(alpha1,beta1); Y = gamma(alpha2,beta2)
# => X/Y approx generalisedbetaprime(alpha1,alpha2,1,beta2/beta1)
# => X/Y approx generalisedbetaprime(kappa1,kappa2,1,theta1/theta2) (kappa=shape,theta=scale)


```


# Bolton cluster

```{r}

# sgeneIncidence = sgll %>% select(FINALID,specimen_date,sgtf_under30CT) %>% left_join(ll2 %>% select(-specimen_date), by=c("FINALID"))
# sgeneIncidence2 = sgeneIncidence %>% filter(LTLA_code %in% areasOfConcern$`Bolton etc`)
# #sgeneIncidence2 %>% group_by(FINALID) %>% filter(n() > 1) %>% arrange(specimen_date) %>% filter(lag(specimen_date)<specimen_date-56) %>% View()
# sgeneIncidence3 = sgeneIncidence2 %>% group_by(FINALID) %>% filter(specimen_date == min(specimen_date)) %>% ungroup()
sgSymptIncid = sgllLoc %>% filter(LTLA_code %in% areasOfConcern$`NW Cluster`) %>% dpc$spim$getLineListIncidence(ageBreaks = c(6,18,45,55,75),subgroup = sgtf_under30CT,filterExpr = asymptomatic_indicator=="N",codeTypes = "LAD") 
#sgSymptIncid = sgllLoc %>% filter(LTLA_code %in% areasOfConcern$`Bolton etc`) %>% dpc$spim$getLineListIncidence(ageBreaks = c(6,18,45,55,75),subgroup = sgtf_under30CT,codeTypes = "LAD") 
tmp = sgSymptIncid %>% dpc$demog$findDemographics()
tmp2 = tmp %>% mutate(code="Bolton etc", name="Bolton etc") %>% ukcovidtools::covidStandardDateGrouping() %>% summarise(value=sum(value), population = sum(population))
tmp2 = tmp2 %>% mutate(subgroup=case_when(
  subgroup==0 ~ "S+",
  subgroup==1 ~ "S-",
  TRUE ~ "equivocal"
))
(tsp$plotIncidenceQuantiles(tmp2 %>% filter(date>"2021-03-01" & ageCat !="unknown"),colour = subgroup,denominatorExpr = population/1000000)+facet_wrap(vars(ageCat))+ylab("daily incidence per 1M")) %>%
  standardPrintOutput::saveThirdPageFigure("~/Dropbox/covid19/sa-variant/BoltonIncidencePer1M")
tmp3 = tmp2 %>% tsp$aggregateAge() %>% tsp$estimateRt(window = 14,quick = FALSE)
tsp$plotRt(tmp3,colour = subgroup,dates =  "2021-03-01", window=14, rtlim = c(0,3)) %>%
  standardPrintOutput::saveThirdPageFigure("~/Dropbox/covid19/sa-variant/BoltonRt")

ggplot(
  sgllLoc %>% #filter(LTLA_code %in% areasOfConcern$`NW Cluster`) %>% 
    filter(specimen_date>"2021-04-01") %>% mutate(sgene=case_when(
  sgtf_under30CT==0 ~ "S+",
  sgtf_under30CT==1 ~ "S-",
  TRUE ~ "equivocal"
)) ,aes(x=age, colour=sgene))+geom_histogram(binwidth = 1)+facet_wrap(vars(sgene), scales = "free")

```

# HERE NEEDS MOVING TO DATA

```{r}

ltlaSgeneCountsAll = sgllLoc %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "LAD", subgroup = sGene)

ltlaToAoc = function(l) l %>% 
  dpc$demog$findDemographics() %>% 
  inner_join(aocDf, by = c("code"="ltlaCode")) %>% 
  select(-code,-name,-name.original) %>%
  rename(name = area) %>%
  mutate(
    name = as.character(name),
    code = name,
    codeType = "AOC"
  ) %>%
  ukcovidtools::covidStandardDateGrouping() %>%
  summarise(value = sum(value), population = sum(population)) %>%
  ungroup()

aocSgeneCounts = bind_rows(
  ltlaToAoc(ltlaSgeneCounts),
  symptomaticPillar2CtrySgeneCounts %>% dpc$demog$findDemographics()
)

aocSgeneCounts %>% filter(subgroup!="equivocal") %>% readr::write_csv("~/Dropbox/covid19/sa-variant/symptomatic-pillar2-cases-by-aoc-and-sgene.csv")

aocSgeneCountsAll = bind_rows(
  ltlaToAoc(ltlaSgeneCountsAll),
  allPillar2CtrySgeneCounts %>% dpc$demog$findDemographics()
)

aocSgeneCountsAll %>% filter(subgroup!="equivocal") %>% readr::write_csv("~/Dropbox/covid19/sa-variant/all-pillar2-cases-by-aoc-and-sgene.csv")

aocDf %>% readr::write_csv("~/Dropbox/covid19/sa-variant/ltla-to-aoc.csv")


  
tsp$plotIncidenceQuantiles(aocSgeneCounts, colour=subgroup, dates="2021-03-01")+facet_wrap(vars(name))+scale_y_continuous(trans="log1p", breaks=c(1,2,5,10,20,50,100,200,500,1000,2000,5000,10000))
tsp$plotIncidenceQuantiles(aocSgeneCountsAll, colour=subgroup, dates="2021-03-01")+facet_wrap(vars(name))+scale_y_continuous(trans="log1p", breaks=c(1,2,5,10,20,50,100,200,500,1000,2000,5000,10000))







# aocSgeneCountsLogGrowth = aocSgeneCounts %>% tsp$logIncidenceStats(smoothingWindow = 21,growthRateWindow = 7, leftSided=TRUE)
# (tsp$plotGrowthRate(aocSgeneCountsLogGrowth %>% 
#               filter(name %in% c("NW Cluster","Bedford etc","Sefton & Lpl")) %>% 
#               filter(subgroup!="equivocal"),colour = subgroup,dates="2021-03-01",rlim=c(-0.25,0.25))+facet_wrap(vars(name))+standardPrintOutput::watermark(lab="provisional")) %>%   
#   standardPrintOutput::saveThirdPageFigure("~/Dropbox/covid19/sa-variant/sGeneAOCHighGrowthLog")
# 
# 
# aocSgeneCountsRtAll = aocSgeneCountsAll %>% tsp$estimateRt(window = 14)
# (tsp$plotRt(aocSgeneCountsRt %>% filter(subgroup!="equivocal"),colour = subgroup,dates="2021-03-01",rtlim=c(0,3.5))+facet_wrap(vars(name))+standardPrintOutput::watermark(lab="provisional")) %>%   
#   standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/sGeneAllAOCRt")


#ggplot(ltlaSgeneCounts, aes(x=date,y=value))
# ltlasSgeneCounts is asymptomatic == "N"
#ltlaSgeneCounts2 = ltlaSgeneCounts %>% filter(subgroup != "equivocal") %>% tsp$logIncidenceStats(growthRateWindow = 7,nocache=TRUE)
ltlaSgeneCounts2 = ltlaSgeneCounts %>% filter(subgroup != "equivocal") %>% tsp$estimateGrowthRate(window = 14, growthRateWindow = 7)
# ltlaSgeneCounts2 = ltlaSgeneCounts2 %>% rename(
#   Growth.windowed.value = Growth.windowed.poisson,
#   Growth.windowed.SE.value = Growth.windowed.SE.poisson
#   ) #etc.

#ltlaSgeneCounts2 = ltlaSgeneCounts %>% filter(subgroup != "equivocal") %>% tsp$estimateGrowthRate(window = 14, growthRateWindow = 7)
```

```{r}
tmp = ltlaSgeneCounts2 %>% select(name,code,date,subgroup,starts_with("Growth"))

tmp2 = tmp %>% 
  filter(subgroup=="S+") %>% inner_join(
    tmp %>% filter(subgroup=="S-"),
    by = c("code","name","date"),
    suffix = c(".pos",".neg")
  ) %>% 
  mutate(
    alpha = 1-sqrt(as.numeric(maxDate-date)/as.numeric(maxDate-minDate))
  ) %>%
  mutate(
    # assume independence between S- and S+ growth rates and that CIs are normally distributes
    Growth.windowed.delta = Growth.windowed.value.pos - Growth.windowed.value.neg,
    Growth.windowed.SE.delta = sqrt(Growth.windowed.SE.value.pos^2+Growth.windowed.SE.value.neg^2),
    Growth.windowed.ProbPos.delta = 1-pnorm(0,mean=Growth.windowed.delta,sd=Growth.windowed.SE.delta),
    Growth.delta = Growth.value.pos - Growth.value.neg,
    Growth.SE.delta = sqrt(Growth.SE.value.pos^2+Growth.SE.value.neg^2),
    Growth.ProbPos.delta = 1-pnorm(0,mean=Growth.delta,sd=Growth.SE.delta)
  ) %>%
  rename(
    `S- Growth Rate` = Growth.value.neg,
    `S+ Growth Rate` = Growth.value.pos
  )

# thanks stackoverflow: https://stats.stackexchange.com/questions/30394/how-to-perform-two-sample-t-tests-in-r-by-inputting-sample-statistics-rather-tha
# m1, m2: the sample means
# s1, s2: the sample standard deviations
# n1, n2: the same sizes
# m0: the null value for the difference in means to be tested for. Default is 0. 
# equal.variance: whether or not to assume equal variance. Default is FALSE. 
differenceStat <- function(m1,m2,s1,s2,n1,n2)
{
    se <- sqrt( (s1^2/n1) + (s2^2/n2) )
    # welch-satterthwaite df
    df <- ( (s1^2/n1 + s2^2/n2)^2 )/( (s1^2/n1)^2/(n1-1) + (s2^2/n2)^2/(n2-1) )
    t <- (m1-m2)/se 
    # difference of means * (1-p-value)
    out = (m1-m2) * (1-2*pt(-abs(t),df))    
    return(out) 
}

diagonals= tibble(
  intercept = c(0.2,0.1,0,-0.1,-0.2),
  slope = 1,
  label = forcats::as_factor(c("-20%","-10%","same","+10%","+20%"))
)


maxDate = max(tmp2$date)-4
minDate = maxDate-28

topNgrowth = tmp2 %>% filter(date==maxDate) %>% mutate(
  diffStat = differenceStat(
    Growth.windowed.value.pos,
    Growth.windowed.value.neg,
    Growth.windowed.SE.value.pos,
    Growth.windowed.SE.value.neg,
    Growth.windowed.Window.value.pos,
    Growth.windowed.Window.value.neg
  )) %>%
  arrange(desc(diffStat)) %>%
  filter(row_number() <= 10) #%>%
  #select(name,code)

# topNgrowth = tmp2 %>% filter(date==maxDate) %>% mutate(
#   #diffStat = Growth.windowed.poisson.pos-Growth.windowed.poisson.neg
#     diffStat = differenceStat(
#       Growth.windowed.poisson.pos,
#       Growth.windowed.poisson.neg,
#       Growth.windowed.SE.poisson.pos,
#       Growth.windowed.SE.poisson.neg,
#       14,14
#     )
#   ) %>% 
#   ungroup() %>%
#   arrange(desc(diffStat)) %>% 
#   filter(row_number() <= 10) %>% 
#   select(name,code)

pathData = tmp2 %>% semi_join(topNgrowth, by="code") %>% 
  filter(date>minDate & date<=maxDate) %>% 
  
#pointData = tmp2 %>% filter(date==maxDate & abs(Growth.windowed.value.pos) >0.01) %>% mutate(alpha = 1)
labelData = tmp2 %>% 
  semi_join(topNgrowth, by="code") %>% 
  filter(date==maxDate) %>% 
  mutate(alpha = 1) %>%
  rename(
    `S- Growth Rate` = Growth.value.neg,
    `S+ Growth Rate` = Growth.value.pos
  )

ggplot(pathData, aes(x=`S- Growth Rate`, y=`S+ Growth Rate`, group=name, alpha=alpha))+
  coord_fixed(xlim=c(-0.2,0.2),ylim=c(-0.2,0.2))+
  geom_hline(yintercept = 0,colour="grey40")+
  geom_vline(xintercept = 0,colour="grey40")+
  geom_abline(data = diagonals, aes(intercept = intercept,slope=slope, colour=label), inherit.aes=FALSE)+
  #geom_abline(slope = 1, intercept = 0.1, colour="#FFB0B0")+
  #geom_abline(slope = 1, intercept = -0.1, colour="#B0FFB0")+
  #geom_abline(slope = 1, intercept = 0.2, colour="#FF8080")+
  #geom_abline(slope = 1, intercept = -0.2, colour="#B0FFB0")+
  #geom_point(data=pointData,colour="#B0B0FF")+
  ggrepel::geom_text_repel(data=labelData,aes(label=name),colour="grey40",min.segment.length = 0,box.padding = 2, 
                           max.overlaps = Inf)+
  geom_path(colour="blue")+
  geom_point(data=labelData,colour="blue")+
  
  guides(alpha="none")+
  xlab("S- growth rate")+
  ylab("S+ growth rate")

# ggplot(pathData, aes(x=Growth.windowed.value.neg, y=Growth.windowed.value.pos, group=name, alpha=alpha))+
#   coord_fixed(xlim=c(-0.15,0.15),ylim=c(-0.15,0.15))+
#   geom_hline(yintercept = 0,colour="grey40")+
#   geom_vline(xintercept = 0,colour="grey40")+
#   geom_abline(data = diagonals, aes(intercept = intercept,slope=slope, colour=label))+
#   #geom_abline(slope = 1, intercept = 0.1, colour="#FFB0B0")+
#   #geom_abline(slope = 1, intercept = -0.1, colour="#B0FFB0")+
#   #geom_abline(slope = 1, intercept = 0.2, colour="#FF8080")+
#   #geom_abline(slope = 1, intercept = -0.2, colour="#B0FFB0")+
#   #geom_point(data=pointData,colour="#B0B0FF")+
#   ggrepel::geom_text_repel(data=labelData,aes(label=name),colour="grey40",min.segment.length = 0,box.padding = 2, 
#                            max.overlaps = Inf)+
#   geom_path(colour="blue")+
#   geom_point(data=labelData,colour="blue")+
#   
#   guides(alpha="none")+
#   xlab("S- growth rate")+
#   ylab("S+ growth rate")

topNgrowth = tmp2 %>% filter(date==maxDate) %>%
  arrange(desc(Growth.windowed.ProbPos.delta*Growth.windowed.delta)) %>%
  filter(row_number() <= 10) %>%
  select(name,code)

(ggplot(tmp2 %>% filter(date>"2020-10-01"),
       aes(
         x=date,
         y=Growth.windowed.delta,
         group=name#,
         #colour=Growth.windowed.ProbPos.delta
         )
        )+geom_line(alpha=0.2)+geom_hline(yintercept = 0,colour="red"))  %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/AllGrowthDelta"))

(ggplot(tmp2 %>% semi_join(topNgrowth) %>% filter(date>"2020-02-01"),
       aes(x=date,
           y=Growth.windowed.delta,
           ymin=Growth.windowed.delta-1.96*Growth.windowed.SE.delta,
           ymax=Growth.windowed.delta+1.96*Growth.windowed.SE.delta,
           group=name)
       )+
  geom_ribbon(alpha=0.2,fill="grey50")+
  geom_line(mapping=aes(colour=name)))  %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/Top10GrowthDelta"))

rollingPercent = allVoc %>% #combinedSpositives %>% 
    filter(!is.na(type)) %>%
    group_by(LTLA_name, type, date=specimen_date) %>%
    summarise(count = n()) %>%
    ungroup() %>%
    tidyr::complete(LTLA_name,type,date,fill=list(count=0)) %>%
    group_by(LTLA_name, type) %>% 
    arrange(date) %>%
    mutate(Roll.count = stats::filter(count,filter=rep(1,28)/28,sides=1)) %>%
    filter(!is.na(Roll.count)) %>% 
    group_by(LTLA_name,date) %>% 
    mutate(binom::binom.confint(Roll.count, sum(Roll.count), conf.level = 0.95, methods = "wilson")) %>% 
    mutate(weight = sum(Roll.count)) %>% 
    ungroup()

growthVPercent = tmp2 %>% inner_join(rollingPercent, by = c("name"="LTLA_name","date"="date")) %>%
  mutate(growthSigHigher = ifelse(Growth.windowed.ProbPos.delta>0.0975,"sig","ns"))

(ggplot(growthVPercent %>% filter(date==max(date)), aes(x=mean,y=Growth.windowed.delta,size=weight))+geom_point(alpha=0.3)+facet_wrap(vars(type),scales="free")+
  #geom_smooth(method = "lm",fullrange=TRUE)+
  coord_cartesian(xlim = c(0,1))) %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/SGeneGrowthDeltaVsPercentSequencing"))

(ggplot(growthVPercent %>% filter(date==max(date)), aes(x=Roll.count,y=Growth.windowed.delta,size=weight))+geom_point(alpha=0.3)+facet_wrap(vars(type))
  #geom_smooth(method = "lm",fullrange=TRUE)+
  ) %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/SGeneGrowthDeltaVsCountSequencing"))

ggplot(growthVPercent %>% filter(type %in% c("B.1.351","B.1.617.1","B.1.617.2","P.1 & P.2") & date==maxDate), aes(x=type,colour=growthSigHigher,y=mean))+geom_boxplot()+scale_y_log10()

```


```{r}
# map = dpc$geog$getMap("LAD19")
# 
# map2 = map %>% inner_join(rollingPercent, by=c("name"="LTLA_name"))
# ggplot(map2 %>% filter(date==maxDate))+geom_sf(aes(fill=mean))+facet_wrap(vars(type))+scale_fill_viridis_c()
#                    
```

```{r}

map3 = map %>% inner_join(tmp2, by="code")
(ggplot(map3 %>% filter(date==maxDate))+geom_sf(aes(fill=Growth.windowed.delta, colour=ifelse(Growth.windowed.ProbPos.delta>0.0975,"sig","ns")),size=0.3)+scale_color_manual(values = c("sig"="white","ns"="black"))+scale_fill_gradient2(high = "yellow", mid="black",low="cyan",limits=c(-0.15,0.15),oob=scales::squish)) %>%
    standardPrintOutput::saveFullPageFigure(paste0("~/Dropbox/covid19/sa-variant/SgeneGrowthDelta-England"))

lapply(names(locations), function(name) {
  limits = locations[[name]]
  xlim = c(limits[2],limits[4])
  ylim = c(limits[1],limits[3])
  (ggplot(map3 %>% filter(date==max(date)))+geom_sf(aes(fill=Growth.windowed.delta))+scale_fill_gradient2(high = "orange",low="blue",limits=c(-0.15,0.15),oob=scales::squish)+coord_sf(xlim=xlim,ylim=ylim)) %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/SgeneGrowthDelta-",name))
})

```

```{r}
topNgrowth %>% inner_join(rollingPercent %>% filter(date==maxDate) %>% mutate(percent =sprintf("%1.1f%%",mean*100)) %>% select(LTLA_name,type,percent)  %>% pivot_wider(values_from = percent, names_from=type), by = c("name"="LTLA_name"))
```


```{r}
ltlaSgeneCounts3 = ltlaSgeneCounts2 %>% dpc$demog$findDemographics()
ltlaSgeneCounts3 = ltlaSgeneCounts3 %>% inner_join(rollingPercent %>% rename(variant=type), by = c("name"="LTLA_name","date"="date"))


tsp$plotGrowthIncidence(ltlaSgeneCounts3 %>% filter(variant == "B.1.617.2") %>% group_by(name, variant,subgroup),plotDates = maxDate,timespan = 28,colour = mean*weight,populationExpr = population,maxSize = 0.6)+facet_wrap(vars(variant,subgroup))+scale_colour_viridis_b()

```


```{r}

sPosll = ll2 %>% select(-specimen_date) %>% inner_join(sgPosUniq %>% filter(sgtf_under30CT == 0) %>% select(FINALID,specimen_date), by="FINALID")

sPosIncid = sPosll %>% dpc$spim$getLineListIncidence(ageBreaks = c(18,35,65,75)) %>% tsp$aggregateGender() %>% dpc$demog$findDemographics()
tsp$plotIncidenceQuantiles(covidTimeseries = sPosIncid %>% filter(name=="England"), dates = "2021-02-01", denominatorExpr = population/1000000, colour=ageCat)+scale_y_continuous(trans="log1p")

tsp$plotRt(covidRtTimeseries = sPosIncid %>% filter(name=="England"), dates = "2021-02-01", colour=ageCat,rtlim = c(0.25,1.5))
```

```{r}
notSNegll = ll %>% semi_join(sgll %>% filter(is.na(sgtf_under30CT) | sgtf_under30CT == 1), by="FINALID")
sNotNegIncid = notSNegll %>% dpc$spim$getLineListIncidence() %>% tsp$aggregateGender() %>% dpc$demog$findDemographics()
tsp$plotIncidenceQuantiles(covidTimeseries = sNotNegIncid, dates = "2021-02-01", denominatorExpr = population/1000000)+facet_wrap(vars(name))+scale_y_continuous(trans="log1p")

tsp$plotRt(covidRtTimeseries = sNotNegIncid, dates = "2021-02-01",rtlim = c(0.25,1.5))+facet_wrap(vars(name))

```

```{r}
sgUniq = sgll %>% group_by(FINALID) %>% 
  filter(CDR_Specimen_Request_SK==min(CDR_Specimen_Request_SK,na.rm=TRUE)) 
llWSgene = ll2 %>% filter(asymptomatic_indicator != "Y") %>% inner_join(sgll %>% filter(!is.na(sgtf_under30CT)) %>% mutate(sGene = ifelse(sgtf_under30CT == 1,"S neg","S pos")) %>% select(FINALID, sGene), by="FINALID")
sGeneIncid = llWSgene %>% dpc$spim$getLineListIncidence(subgroup = sGene) %>% dpc$demog$findDemographics()
tsp$plotIncidenceQuantiles(covidTimeseries = sGeneIncid, dates = "2021-02-01", colour = subgroup, denominatorExpr = population/1000000)+facet_wrap(vars(name))+scale_y_continuous(trans="log1p")

sRT14 = sGeneIncid %>% filter(name != "Unknown (England)") %>% tsp$estimateRt(quick = TRUE,window = 14)
tsp$plotRt(covidRtTimeseries = sRT14, dates = "2021-02-01",rtlim = c(0.25,1.5), colour = subgroup)+facet_wrap(vars(name))

```


# Bolton cluster

```{r}

# sgeneIncidence = sgll %>% select(FINALID,specimen_date,sgtf_under30CT) %>% left_join(ll2 %>% select(-specimen_date), by=c("FINALID"))
# sgeneIncidence2 = sgeneIncidence %>% filter(LTLA_code %in% areasOfConcern$`Bolton etc`)
# #sgeneIncidence2 %>% group_by(FINALID) %>% filter(n() > 1) %>% arrange(specimen_date) %>% filter(lag(specimen_date)<specimen_date-56) %>% View()
# sgeneIncidence3 = sgeneIncidence2 %>% group_by(FINALID) %>% filter(specimen_date == min(specimen_date)) %>% ungroup()
sgSymptIncid = sgllLoc %>% filter(LTLA_code %in% areasOfConcern$`NW Cluster`) %>% dpc$spim$getLineListIncidence(ageBreaks = c(6,18,45,55,75),subgroup = sgtf_under30CT,filterExpr = asymptomatic_indicator=="N",codeTypes = "LAD") 
#sgSymptIncid = sgllLoc %>% filter(LTLA_code %in% areasOfConcern$`Bolton etc`) %>% dpc$spim$getLineListIncidence(ageBreaks = c(6,18,45,55,75),subgroup = sgtf_under30CT,codeTypes = "LAD") 
tmp = sgSymptIncid %>% dpc$demog$findDemographics()
tmp2 = tmp %>% mutate(code="Bolton etc", name="Bolton etc") %>% ukcovidtools::covidStandardDateGrouping() %>% summarise(value=sum(value), population = sum(population))
tmp2 = tmp2 %>% mutate(subgroup=case_when(
  subgroup==0 ~ "S+",
  subgroup==1 ~ "S-",
  TRUE ~ "equivocal"
))
(tsp$plotIncidenceQuantiles(tmp2 %>% filter(date>"2021-03-01" & ageCat !="unknown"),colour = subgroup,denominatorExpr = population/1000000)+facet_wrap(vars(ageCat))+ylab("daily incidence per 1M")) %>%
  standardPrintOutput::saveThirdPageFigure("~/Dropbox/covid19/sa-variant/BoltonIncidencePer1M")
tmp3 = tmp2 %>% tsp$aggregateAge() %>% tsp$estimateRt(window = 14,quick = FALSE)
tsp$plotRt(tmp3,colour = subgroup,dates =  "2021-03-01", window=14, rtlim = c(0,3)) %>%
  standardPrintOutput::saveThirdPageFigure("~/Dropbox/covid19/sa-variant/BoltonRt")

ggplot(
  sgllLoc %>% #filter(LTLA_code %in% areasOfConcern$`NW Cluster`) %>% 
    filter(specimen_date>"2021-04-01") %>% mutate(sgene=case_when(
  sgtf_under30CT==0 ~ "S+",
  sgtf_under30CT==1 ~ "S-",
  TRUE ~ "equivocal"
)) ,aes(x=age, colour=sgene))+geom_histogram(binwidth = 1)+facet_wrap(vars(sgene), scales = "free")

```
```{r}



```

```{r}

ltlaSgeneCountsAll = sgllLoc %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "LAD", subgroup = sGene)

ltlaToAoc = function(l) l %>% 
  dpc$demog$findDemographics() %>% 
  inner_join(aocDf, by = c("code"="ltlaCode")) %>% 
  select(-code,-name,-name.original) %>%
  rename(name = area) %>%
  mutate(
    name = as.character(name),
    code = name,
    codeType = "AOC"
  ) %>%
  ukcovidtools::covidStandardDateGrouping() %>%
  summarise(value = sum(value), population = sum(population)) %>%
  ungroup()

aocSgeneCounts = bind_rows(
  ltlaToAoc(ltlaSgeneCounts),
  symptomaticPillar2CtrySgeneCounts %>% dpc$demog$findDemographics()
)

aocSgeneCounts %>% filter(subgroup!="equivocal") %>% readr::write_csv("~/Dropbox/covid19/sa-variant/symptomatic-pillar2-cases-by-aoc-and-sgene.csv")

aocSgeneCountsAll = bind_rows(
  ltlaToAoc(ltlaSgeneCountsAll),
  allPillar2CtrySgeneCounts %>% dpc$demog$findDemographics()
)

aocSgeneCountsAll %>% filter(subgroup!="equivocal") %>% readr::write_csv("~/Dropbox/covid19/sa-variant/all-pillar2-cases-by-aoc-and-sgene.csv")

aocDf %>% readr::write_csv("~/Dropbox/covid19/sa-variant/ltla-to-aoc.csv")


  
tsp$plotIncidenceQuantiles(aocSgeneCounts, colour=subgroup, dates="2021-03-01")+facet_wrap(vars(name))+scale_y_continuous(trans="log1p", breaks=c(1,2,5,10,20,50,100,200,500,1000,2000,5000,10000))
tsp$plotIncidenceQuantiles(aocSgeneCountsAll, colour=subgroup, dates="2021-03-01")+facet_wrap(vars(name))+scale_y_continuous(trans="log1p", breaks=c(1,2,5,10,20,50,100,200,500,1000,2000,5000,10000))



clusters = list(
  "CommunityIndia2" = c("NW Cluster","Bedford etc","Sefton & Lpl"),
  "MixedIndia2" = c("Leicester","Nottingham","West London"),
  "MixedSA" = c("East London"),
  "Reference" = c("England")
)

aocSgeneCountsGrowth = aocSgeneCounts %>% tsp$estimateGrowthRate(window = 56,growthRateWindow = 7)

lapply(names(clusters),function(clustName) {
  (
    tsp$plotGrowthRate(
      aocSgeneCountsGrowth %>% filter(name %in% clusters[[clustName]]) %>% filter(subgroup!="equivocal"),
      colour = subgroup,dates="2021-03-01",
      rlim=c(-0.25,0.25),
      growthVar = Growth.poisson,
      growthSEVar = Growth.SE.poisson,
    )+facet_wrap(vars(name))+standardPrintOutput::watermark(lab="provisional")+ylab("r - sliding 8 week window")
  ) %>%   
  standardPrintOutput::saveThirdPageFigure(paste0("~/Dropbox/covid19/sa-variant/sGeneAOCHighGrowth-",clustName))
})

aocSgeneCountsRt = aocSgeneCounts %>% tsp$estimateRt(window = 14)
lapply(names(clusters),function(clustName) {
  (
    tsp$plotRt(
      aocSgeneCountsRt %>% filter(name %in% clusters[[clustName]]) %>% filter(subgroup!="equivocal"),
      colour = subgroup,
      dates="2021-03-01",
      rtlim=c(0,3.5)
    )+
    facet_wrap(vars(name))+
    standardPrintOutput::watermark(lab="provisional")) %>%   
  standardPrintOutput::saveThirdPageFigure(paste0("~/Dropbox/covid19/sa-variant/sGeneAOCHighRt-",clustName))
})




# aocSgeneCountsLogGrowth = aocSgeneCounts %>% tsp$logIncidenceStats(smoothingWindow = 21,growthRateWindow = 7, leftSided=TRUE)
# (tsp$plotGrowthRate(aocSgeneCountsLogGrowth %>% 
#               filter(name %in% c("NW Cluster","Bedford etc","Sefton & Lpl")) %>% 
#               filter(subgroup!="equivocal"),colour = subgroup,dates="2021-03-01",rlim=c(-0.25,0.25))+facet_wrap(vars(name))+standardPrintOutput::watermark(lab="provisional")) %>%   
#   standardPrintOutput::saveThirdPageFigure("~/Dropbox/covid19/sa-variant/sGeneAOCHighGrowthLog")
# 
# 
# aocSgeneCountsRtAll = aocSgeneCountsAll %>% tsp$estimateRt(window = 14)
# (tsp$plotRt(aocSgeneCountsRt %>% filter(subgroup!="equivocal"),colour = subgroup,dates="2021-03-01",rtlim=c(0,3.5))+facet_wrap(vars(name))+standardPrintOutput::watermark(lab="provisional")) %>%   
#   standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/sGeneAllAOCRt")


#ggplot(ltlaSgeneCounts, aes(x=date,y=value))
# ltlasSgeneCounts is asymptomatic == "N"
#ltlaSgeneCounts2 = ltlaSgeneCounts %>% filter(subgroup != "equivocal") %>% tsp$logIncidenceStats(growthRateWindow = 7,nocache=TRUE)
ltlaSgeneCounts2 = ltlaSgeneCounts %>% filter(subgroup != "equivocal") %>% tsp$estimateGrowthRate(window = 14, growthRateWindow = 7)
# ltlaSgeneCounts2 = ltlaSgeneCounts2 %>% rename(
#   Growth.windowed.value = Growth.windowed.poisson,
#   Growth.windowed.SE.value = Growth.windowed.SE.poisson
#   ) #etc.

#ltlaSgeneCounts2 = ltlaSgeneCounts %>% filter(subgroup != "equivocal") %>% tsp$estimateGrowthRate(window = 14, growthRateWindow = 7)
```

```{r}
tmp = ltlaSgeneCounts2 %>% select(name,code,date,subgroup,starts_with("Growth"))

tmp2 = tmp %>% 
  filter(subgroup=="S+") %>% inner_join(
    tmp %>% filter(subgroup=="S-"),
    by = c("code","name","date"),
    suffix = c(".pos",".neg")
  ) %>% 
  mutate(
    alpha = 1-sqrt(as.numeric(maxDate-date)/as.numeric(maxDate-minDate))
  ) %>%
  mutate(
    # assume independence between S- and S+ growth rates and that CIs are normally distributes
    Growth.windowed.delta = Growth.windowed.value.pos - Growth.windowed.value.neg,
    Growth.windowed.SE.delta = sqrt(Growth.windowed.SE.value.pos^2+Growth.windowed.SE.value.neg^2),
    Growth.windowed.ProbPos.delta = 1-pnorm(0,mean=Growth.windowed.delta,sd=Growth.windowed.SE.delta),
    Growth.delta = Growth.value.pos - Growth.value.neg,
    Growth.SE.delta = sqrt(Growth.SE.value.pos^2+Growth.SE.value.neg^2),
    Growth.ProbPos.delta = 1-pnorm(0,mean=Growth.delta,sd=Growth.SE.delta)
  ) %>%
  rename(
    `S- Growth Rate` = Growth.value.neg,
    `S+ Growth Rate` = Growth.value.pos
  )

# thanks stackoverflow: https://stats.stackexchange.com/questions/30394/how-to-perform-two-sample-t-tests-in-r-by-inputting-sample-statistics-rather-tha
# m1, m2: the sample means
# s1, s2: the sample standard deviations
# n1, n2: the same sizes
# m0: the null value for the difference in means to be tested for. Default is 0. 
# equal.variance: whether or not to assume equal variance. Default is FALSE. 
differenceStat <- function(m1,m2,s1,s2,n1,n2)
{
    se <- sqrt( (s1^2/n1) + (s2^2/n2) )
    # welch-satterthwaite df
    df <- ( (s1^2/n1 + s2^2/n2)^2 )/( (s1^2/n1)^2/(n1-1) + (s2^2/n2)^2/(n2-1) )
    t <- (m1-m2)/se 
    # difference of means * (1-p-value)
    out = (m1-m2) * (1-2*pt(-abs(t),df))    
    return(out) 
}

diagonals= tibble(
  intercept = c(0.2,0.1,0,-0.1,-0.2),
  slope = 1,
  label = forcats::as_factor(c("-20%","-10%","same","+10%","+20%"))
)


maxDate = max(tmp2$date)-4
minDate = maxDate-28

topNgrowth = tmp2 %>% filter(date==maxDate) %>% mutate(
  diffStat = differenceStat(
    Growth.windowed.value.pos,
    Growth.windowed.value.neg,
    Growth.windowed.SE.value.pos,
    Growth.windowed.SE.value.neg,
    Growth.windowed.Window.value.pos,
    Growth.windowed.Window.value.neg
  )) %>%
  arrange(desc(diffStat)) %>%
  filter(row_number() <= 10) #%>%
  #select(name,code)

# topNgrowth = tmp2 %>% filter(date==maxDate) %>% mutate(
#   #diffStat = Growth.windowed.poisson.pos-Growth.windowed.poisson.neg
#     diffStat = differenceStat(
#       Growth.windowed.poisson.pos,
#       Growth.windowed.poisson.neg,
#       Growth.windowed.SE.poisson.pos,
#       Growth.windowed.SE.poisson.neg,
#       14,14
#     )
#   ) %>% 
#   ungroup() %>%
#   arrange(desc(diffStat)) %>% 
#   filter(row_number() <= 10) %>% 
#   select(name,code)

pathData = tmp2 %>% semi_join(topNgrowth, by="code") %>% 
  filter(date>minDate & date<=maxDate) %>% 
  
#pointData = tmp2 %>% filter(date==maxDate & abs(Growth.windowed.value.pos) >0.01) %>% mutate(alpha = 1)
labelData = tmp2 %>% 
  semi_join(topNgrowth, by="code") %>% 
  filter(date==maxDate) %>% 
  mutate(alpha = 1) %>%
  rename(
    `S- Growth Rate` = Growth.value.neg,
    `S+ Growth Rate` = Growth.value.pos
  )

ggplot(pathData, aes(x=`S- Growth Rate`, y=`S+ Growth Rate`, group=name, alpha=alpha))+
  coord_fixed(xlim=c(-0.2,0.2),ylim=c(-0.2,0.2))+
  geom_hline(yintercept = 0,colour="grey40")+
  geom_vline(xintercept = 0,colour="grey40")+
  geom_abline(data = diagonals, aes(intercept = intercept,slope=slope, colour=label), inherit.aes=FALSE)+
  #geom_abline(slope = 1, intercept = 0.1, colour="#FFB0B0")+
  #geom_abline(slope = 1, intercept = -0.1, colour="#B0FFB0")+
  #geom_abline(slope = 1, intercept = 0.2, colour="#FF8080")+
  #geom_abline(slope = 1, intercept = -0.2, colour="#B0FFB0")+
  #geom_point(data=pointData,colour="#B0B0FF")+
  ggrepel::geom_text_repel(data=labelData,aes(label=name),colour="grey40",min.segment.length = 0,box.padding = 2, 
                           max.overlaps = Inf)+
  geom_path(colour="blue")+
  geom_point(data=labelData,colour="blue")+
  
  guides(alpha="none")+
  xlab("S- growth rate")+
  ylab("S+ growth rate")

# ggplot(pathData, aes(x=Growth.windowed.value.neg, y=Growth.windowed.value.pos, group=name, alpha=alpha))+
#   coord_fixed(xlim=c(-0.15,0.15),ylim=c(-0.15,0.15))+
#   geom_hline(yintercept = 0,colour="grey40")+
#   geom_vline(xintercept = 0,colour="grey40")+
#   geom_abline(data = diagonals, aes(intercept = intercept,slope=slope, colour=label))+
#   #geom_abline(slope = 1, intercept = 0.1, colour="#FFB0B0")+
#   #geom_abline(slope = 1, intercept = -0.1, colour="#B0FFB0")+
#   #geom_abline(slope = 1, intercept = 0.2, colour="#FF8080")+
#   #geom_abline(slope = 1, intercept = -0.2, colour="#B0FFB0")+
#   #geom_point(data=pointData,colour="#B0B0FF")+
#   ggrepel::geom_text_repel(data=labelData,aes(label=name),colour="grey40",min.segment.length = 0,box.padding = 2, 
#                            max.overlaps = Inf)+
#   geom_path(colour="blue")+
#   geom_point(data=labelData,colour="blue")+
#   
#   guides(alpha="none")+
#   xlab("S- growth rate")+
#   ylab("S+ growth rate")

topNgrowth = tmp2 %>% filter(date==maxDate) %>%
  arrange(desc(Growth.windowed.ProbPos.delta*Growth.windowed.delta)) %>%
  filter(row_number() <= 10) %>%
  select(name,code)

(ggplot(tmp2 %>% filter(date>"2020-10-01"),
       aes(
         x=date,
         y=Growth.windowed.delta,
         group=name#,
         #colour=Growth.windowed.ProbPos.delta
         )
        )+geom_line(alpha=0.2)+geom_hline(yintercept = 0,colour="red"))  %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/AllGrowthDelta"))

(ggplot(tmp2 %>% semi_join(topNgrowth) %>% filter(date>"2020-02-01"),
       aes(x=date,
           y=Growth.windowed.delta,
           ymin=Growth.windowed.delta-1.96*Growth.windowed.SE.delta,
           ymax=Growth.windowed.delta+1.96*Growth.windowed.SE.delta,
           group=name)
       )+
  geom_ribbon(alpha=0.2,fill="grey50")+
  geom_line(mapping=aes(colour=name)))  %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/Top10GrowthDelta"))

rollingPercent = allVoc %>% #combinedSpositives %>% 
    filter(!is.na(type)) %>%
    group_by(LTLA_name, type, date=specimen_date) %>%
    summarise(count = n()) %>%
    ungroup() %>%
    tidyr::complete(LTLA_name,type,date,fill=list(count=0)) %>%
    group_by(LTLA_name, type) %>% 
    arrange(date) %>%
    mutate(Roll.count = stats::filter(count,filter=rep(1,28)/28,sides=1)) %>%
    filter(!is.na(Roll.count)) %>% 
    group_by(LTLA_name,date) %>% 
    mutate(binom::binom.confint(Roll.count, sum(Roll.count), conf.level = 0.95, methods = "wilson")) %>% 
    mutate(weight = sum(Roll.count)) %>% 
    ungroup()

growthVPercent = tmp2 %>% inner_join(rollingPercent, by = c("name"="LTLA_name","date"="date")) %>%
  mutate(growthSigHigher = ifelse(Growth.windowed.ProbPos.delta>0.0975,"sig","ns"))

(ggplot(growthVPercent %>% filter(date==max(date)), aes(x=mean,y=Growth.windowed.delta,size=weight))+geom_point(alpha=0.3)+facet_wrap(vars(type),scales="free")+
  #geom_smooth(method = "lm",fullrange=TRUE)+
  coord_cartesian(xlim = c(0,1))) %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/SGeneGrowthDeltaVsPercentSequencing"))

(ggplot(growthVPercent %>% filter(date==max(date)), aes(x=Roll.count,y=Growth.windowed.delta,size=weight))+geom_point(alpha=0.3)+facet_wrap(vars(type))
  #geom_smooth(method = "lm",fullrange=TRUE)+
  ) %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/SGeneGrowthDeltaVsCountSequencing"))

ggplot(growthVPercent %>% filter(type %in% c("B.1.351","B.1.617.1","B.1.617.2","P.1 & P.2") & date==maxDate), aes(x=type,colour=growthSigHigher,y=mean))+geom_boxplot()+scale_y_log10()

```


```{r}
# map = dpc$geog$getMap("LAD19")
# 
# map2 = map %>% inner_join(rollingPercent, by=c("name"="LTLA_name"))
# ggplot(map2 %>% filter(date==maxDate))+geom_sf(aes(fill=mean))+facet_wrap(vars(type))+scale_fill_viridis_c()
#                    
```

```{r}

map3 = map %>% inner_join(tmp2, by="code")
(ggplot(map3 %>% filter(date==maxDate))+geom_sf(aes(fill=Growth.windowed.delta, colour=ifelse(Growth.windowed.ProbPos.delta>0.0975,"sig","ns")),size=0.3)+scale_color_manual(values = c("sig"="white","ns"="black"))+scale_fill_gradient2(high = "yellow", mid="black",low="cyan",limits=c(-0.15,0.15),oob=scales::squish)) %>%
    standardPrintOutput::saveFullPageFigure(paste0("~/Dropbox/covid19/sa-variant/SgeneGrowthDelta-England"))

lapply(names(locations), function(name) {
  limits = locations[[name]]
  xlim = c(limits[2],limits[4])
  ylim = c(limits[1],limits[3])
  (ggplot(map3 %>% filter(date==max(date)))+geom_sf(aes(fill=Growth.windowed.delta))+scale_fill_gradient2(high = "orange",low="blue",limits=c(-0.15,0.15),oob=scales::squish)+coord_sf(xlim=xlim,ylim=ylim)) %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/SgeneGrowthDelta-",name))
})

```

```{r}
topNgrowth %>% inner_join(rollingPercent %>% filter(date==maxDate) %>% mutate(percent =sprintf("%1.1f%%",mean*100)) %>% select(LTLA_name,type,percent)  %>% pivot_wider(values_from = percent, names_from=type), by = c("name"="LTLA_name"))
```


```{r}
ltlaSgeneCounts3 = ltlaSgeneCounts2 %>% dpc$demog$findDemographics()
ltlaSgeneCounts3 = ltlaSgeneCounts3 %>% inner_join(rollingPercent %>% rename(variant=type), by = c("name"="LTLA_name","date"="date"))


tsp$plotGrowthIncidence(ltlaSgeneCounts3 %>% filter(variant == "B.1.617.2") %>% group_by(name, variant,subgroup),plotDates = maxDate,timespan = 28,colour = mean*weight,populationExpr = population,maxSize = 0.6)+facet_wrap(vars(variant,subgroup))+scale_colour_viridis_b()

```