---
title: "B.1.251 sequences and S-gene positive pillar 2 cases"
author: "Rob Challen"
date: '`r format(Sys.Date(), "%d-%m-%Y")`'
output: 
  pdf_document :
    fig_caption: yes
header-includes:
  \usepackage{float}
  \floatplacement{figure}{H}
knit: (function(inputFile, encoding,...) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "~/Dropbox/covid19/sa-variant/")})
fig_width: 7
fig_height: 5
out.width: "100%"
---


```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo = FALSE,
  warning = FALSE,
  message = FALSE
)

devtools::load_all("~/Git/standard-print-output/")
devtools::load_all("~/Git/uk-covid-datatools/")

library(tidyverse)
library(patchwork)
library(rgdal)
library(ggplot2)
library(ggspatial)
library(rgeos)
library(maptools)
library(patchwork)
library(sp)
library(sf)

ggplot2::theme_set(standardPrintOutput::defaultFigureLayout())
standardPrintOutput::setDefaults()
ukcovidtools::setup()
ukcovidtools::reload()
```

```{r eval=FALSE}
devtools::load_all("~/Git/uk-covid-datatools/")
ukcovidtools::reload()

setwd("~/Git/uk-covid-datatools/")
if (!exists("combinedSpositives")) source("./vignettes/b-1-351-s-pos-data.R")
```

# Data set for S-gene growth rates

```{r}
ll2 = ll %>% dpc$spim$augmentLineListWithLSOA()
## Generate the s-gene line list with S-gene status and location

sgllLoc = sgllEra %>% select(FINALID, specimen_date = earliest_specimen_date, latest_specimen_date, sgtf_under30CT, infection=eraIndex) %>%
  inner_join(ll2, by=c("FINALID"), suffix=c("",".ll")) %>% mutate(
    asymptomatic_indicator = ifelse(specimen_date==specimen_date.ll,asymptomatic_indicator,NA_character_),
    pillar_2_testingkit = ifelse(specimen_date==specimen_date.ll,pillar_2_testingkit,NA_character_)
      #ifelse(specimen_date<=specimen_date.ll & latest_specimen_date>=specimen_date.ll, asymptomatic_indicator,NA_character_),
  ) %>% select(-cat,-pillar_2_testingkit, -testcentreid, -case_category) %>%
  mutate(sGene = case_when(
    sgtf_under30CT == 0 ~ "S+",
    sgtf_under30CT == 1 ~ "S-",
    TRUE ~ "equivocal",
  ))

ltlaSgeneCounts = sgllLoc %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "LAD", filterExpr = asymptomatic_indicator=="N",subgroup = sGene)

symptomaticPillar2CtrySgeneCounts = sgllLoc %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "CTRY", filterExpr = asymptomatic_indicator=="N",subgroup = sGene)
allPillar2CtrySgeneCounts = sgllLoc %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "CTRY",subgroup = sGene)

allPillar2CtrySgeneCounts %>% readr::write_csv(paste0("~/Dropbox/covid19/sa-variant/eng-case-counts-by-sgene-",Sys.Date(),".csv")) 
symptomaticPillar2CtrySgeneCounts %>% readr::write_csv(paste0("~/Dropbox/covid19/sa-variant/eng-symptomatic-case-counts-by-sgene-",Sys.Date(),".csv"))
```

```{r}

sPosll = ll2 %>% select(-specimen_date) %>% inner_join(sgPosUniq %>% filter(sgtf_under30CT == 0) %>% select(FINALID,specimen_date), by="FINALID")

sPosIncid = sPosll %>% dpc$spim$getLineListIncidence(ageBreaks = c(18,35,65,75)) %>% tsp$aggregateGender() %>% dpc$demog$findDemographics()
tsp$plotIncidenceQuantiles(covidTimeseries = sPosIncid %>% filter(name=="England"), dates = "2021-02-01", denominatorExpr = population/1000000, colour=ageCat)+scale_y_continuous(trans="log1p")

tsp$plotRt(covidRtTimeseries = sPosIncid %>% filter(name=="England"), dates = "2021-02-01", colour=ageCat,rtlim = c(0.25,1.5))
```

```{r}
notSNegll = ll %>% semi_join(sgll %>% filter(is.na(sgtf_under30CT) | sgtf_under30CT == 1), by="FINALID")
sNotNegIncid = notSNegll %>% dpc$spim$getLineListIncidence() %>% tsp$aggregateGender() %>% dpc$demog$findDemographics()
tsp$plotIncidenceQuantiles(covidTimeseries = sNotNegIncid, dates = "2021-02-01", denominatorExpr = population/1000000)+facet_wrap(vars(name))+scale_y_continuous(trans="log1p")

tsp$plotRt(covidRtTimeseries = sNotNegIncid, dates = "2021-02-01",rtlim = c(0.25,1.5))+facet_wrap(vars(name))

```

```{r}
sgUniq = sgll %>% group_by(FINALID) %>% 
  filter(CDR_Specimen_Request_SK==min(CDR_Specimen_Request_SK,na.rm=TRUE)) 
llWSgene = ll2 %>% filter(asymptomatic_indicator != "Y") %>% inner_join(sgll %>% filter(!is.na(sgtf_under30CT)) %>% mutate(sGene = ifelse(sgtf_under30CT == 1,"S neg","S pos")) %>% select(FINALID, sGene), by="FINALID")
sGeneIncid = llWSgene %>% dpc$spim$getLineListIncidence(subgroup = sGene) %>% dpc$demog$findDemographics()
tsp$plotIncidenceQuantiles(covidTimeseries = sGeneIncid, dates = "2021-02-01", colour = subgroup, denominatorExpr = population/1000000)+facet_wrap(vars(name))+scale_y_continuous(trans="log1p")

sRT14 = sGeneIncid %>% filter(name != "Unknown (England)") %>% tsp$estimateRt(quick = TRUE,window = 14)
tsp$plotRt(covidRtTimeseries = sRT14, dates = "2021-02-01",rtlim = c(0.25,1.5), colour = subgroup)+facet_wrap(vars(name))

```


# Bolton cluster

```{r}

# sgeneIncidence = sgll %>% select(FINALID,specimen_date,sgtf_under30CT) %>% left_join(ll2 %>% select(-specimen_date), by=c("FINALID"))
# sgeneIncidence2 = sgeneIncidence %>% filter(LTLA_code %in% areasOfConcern$`Bolton etc`)
# #sgeneIncidence2 %>% group_by(FINALID) %>% filter(n() > 1) %>% arrange(specimen_date) %>% filter(lag(specimen_date)<specimen_date-56) %>% View()
# sgeneIncidence3 = sgeneIncidence2 %>% group_by(FINALID) %>% filter(specimen_date == min(specimen_date)) %>% ungroup()
sgSymptIncid = sgllLoc %>% filter(LTLA_code %in% areasOfConcern$`NW Cluster`) %>% dpc$spim$getLineListIncidence(ageBreaks = c(6,18,45,55,75),subgroup = sgtf_under30CT,filterExpr = asymptomatic_indicator=="N",codeTypes = "LAD") 
#sgSymptIncid = sgllLoc %>% filter(LTLA_code %in% areasOfConcern$`Bolton etc`) %>% dpc$spim$getLineListIncidence(ageBreaks = c(6,18,45,55,75),subgroup = sgtf_under30CT,codeTypes = "LAD") 
tmp = sgSymptIncid %>% dpc$demog$findDemographics()
tmp2 = tmp %>% mutate(code="Bolton etc", name="Bolton etc") %>% ukcovidtools::covidStandardDateGrouping() %>% summarise(value=sum(value), population = sum(population))
tmp2 = tmp2 %>% mutate(subgroup=case_when(
  subgroup==0 ~ "S+",
  subgroup==1 ~ "S-",
  TRUE ~ "equivocal"
))
(tsp$plotIncidenceQuantiles(tmp2 %>% filter(date>"2021-03-01" & ageCat !="unknown"),colour = subgroup,denominatorExpr = population/1000000)+facet_wrap(vars(ageCat))+ylab("daily incidence per 1M")) %>%
  standardPrintOutput::saveThirdPageFigure("~/Dropbox/covid19/sa-variant/BoltonIncidencePer1M")
tmp3 = tmp2 %>% tsp$aggregateAge() %>% tsp$estimateRt(window = 14,quick = FALSE)
tsp$plotRt(tmp3,colour = subgroup,dates =  "2021-03-01", window=14, rtlim = c(0,3)) %>%
  standardPrintOutput::saveThirdPageFigure("~/Dropbox/covid19/sa-variant/BoltonRt")

ggplot(
  sgllLoc %>% #filter(LTLA_code %in% areasOfConcern$`NW Cluster`) %>% 
    filter(specimen_date>"2021-04-01") %>% mutate(sgene=case_when(
  sgtf_under30CT==0 ~ "S+",
  sgtf_under30CT==1 ~ "S-",
  TRUE ~ "equivocal"
)) ,aes(x=age, colour=sgene))+geom_histogram(binwidth = 1)+facet_wrap(vars(sgene), scales = "free")

```

# HERE NEEDS MOVING TO DATA

```{r}

ltlaSgeneCountsAll = sgllLoc %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "LAD", subgroup = sGene)

ltlaToAoc = function(l) l %>% 
  dpc$demog$findDemographics() %>% 
  inner_join(aocDf, by = c("code"="ltlaCode")) %>% 
  select(-code,-name,-name.original) %>%
  rename(name = area) %>%
  mutate(
    name = as.character(name),
    code = name,
    codeType = "AOC"
  ) %>%
  ukcovidtools::covidStandardDateGrouping() %>%
  summarise(value = sum(value), population = sum(population)) %>%
  ungroup()

aocSgeneCounts = bind_rows(
  ltlaToAoc(ltlaSgeneCounts),
  symptomaticPillar2CtrySgeneCounts %>% dpc$demog$findDemographics()
)

aocSgeneCounts %>% filter(subgroup!="equivocal") %>% readr::write_csv("~/Dropbox/covid19/sa-variant/symptomatic-pillar2-cases-by-aoc-and-sgene.csv")

aocSgeneCountsAll = bind_rows(
  ltlaToAoc(ltlaSgeneCountsAll),
  allPillar2CtrySgeneCounts %>% dpc$demog$findDemographics()
)

aocSgeneCountsAll %>% filter(subgroup!="equivocal") %>% readr::write_csv("~/Dropbox/covid19/sa-variant/all-pillar2-cases-by-aoc-and-sgene.csv")

aocDf %>% readr::write_csv("~/Dropbox/covid19/sa-variant/ltla-to-aoc.csv")


  
tsp$plotIncidenceQuantiles(aocSgeneCounts, colour=subgroup, dates="2021-03-01")+facet_wrap(vars(name))+scale_y_continuous(trans="log1p", breaks=c(1,2,5,10,20,50,100,200,500,1000,2000,5000,10000))
tsp$plotIncidenceQuantiles(aocSgeneCountsAll, colour=subgroup, dates="2021-03-01")+facet_wrap(vars(name))+scale_y_continuous(trans="log1p", breaks=c(1,2,5,10,20,50,100,200,500,1000,2000,5000,10000))



clusters = list(
  "CommunityIndia2" = c("NW Cluster","Bedford etc","Sefton & Lpl"),
  "MixedIndia2" = c("Leicester","Nottingham","West London"),
  "MixedSA" = c("East London"),
  "Reference" = c("England")
)

aocSgeneCountsGrowth = aocSgeneCounts %>% tsp$estimateGrowthRate(window = 56,growthRateWindow = 7)

lapply(names(clusters),function(clustName) {
  (
    tsp$plotGrowthRate(
      aocSgeneCountsGrowth %>% filter(name %in% clusters[[clustName]]) %>% filter(subgroup!="equivocal"),
      colour = subgroup,dates="2021-03-01",
      rlim=c(-0.25,0.25),
      growthVar = Growth.poisson,
      growthSEVar = Growth.SE.poisson,
    )+facet_wrap(vars(name))+standardPrintOutput::watermark(lab="provisional")+ylab("r - sliding 8 week window")
  ) %>%   
  standardPrintOutput::saveThirdPageFigure(paste0("~/Dropbox/covid19/sa-variant/sGeneAOCHighGrowth-",clustName))
})

aocSgeneCountsRt = aocSgeneCounts %>% tsp$estimateRt(window = 14)
lapply(names(clusters),function(clustName) {
  (
    tsp$plotRt(
      aocSgeneCountsRt %>% filter(name %in% clusters[[clustName]]) %>% filter(subgroup!="equivocal"),
      colour = subgroup,
      dates="2021-03-01",
      rtlim=c(0,3.5)
    )+
    facet_wrap(vars(name))+
    standardPrintOutput::watermark(lab="provisional")) %>%   
  standardPrintOutput::saveThirdPageFigure(paste0("~/Dropbox/covid19/sa-variant/sGeneAOCHighRt-",clustName))
})




# aocSgeneCountsLogGrowth = aocSgeneCounts %>% tsp$logIncidenceStats(smoothingWindow = 21,growthRateWindow = 7, leftSided=TRUE)
# (tsp$plotGrowthRate(aocSgeneCountsLogGrowth %>% 
#               filter(name %in% c("NW Cluster","Bedford etc","Sefton & Lpl")) %>% 
#               filter(subgroup!="equivocal"),colour = subgroup,dates="2021-03-01",rlim=c(-0.25,0.25))+facet_wrap(vars(name))+standardPrintOutput::watermark(lab="provisional")) %>%   
#   standardPrintOutput::saveThirdPageFigure("~/Dropbox/covid19/sa-variant/sGeneAOCHighGrowthLog")
# 
# 
# aocSgeneCountsRtAll = aocSgeneCountsAll %>% tsp$estimateRt(window = 14)
# (tsp$plotRt(aocSgeneCountsRt %>% filter(subgroup!="equivocal"),colour = subgroup,dates="2021-03-01",rtlim=c(0,3.5))+facet_wrap(vars(name))+standardPrintOutput::watermark(lab="provisional")) %>%   
#   standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/sGeneAllAOCRt")


#ggplot(ltlaSgeneCounts, aes(x=date,y=value))
# ltlasSgeneCounts is asymptomatic == "N"
#ltlaSgeneCounts2 = ltlaSgeneCounts %>% filter(subgroup != "equivocal") %>% tsp$logIncidenceStats(growthRateWindow = 7,nocache=TRUE)
ltlaSgeneCounts2 = ltlaSgeneCounts %>% filter(subgroup != "equivocal") %>% tsp$estimateGrowthRate(window = 14, growthRateWindow = 7)
# ltlaSgeneCounts2 = ltlaSgeneCounts2 %>% rename(
#   Growth.windowed.value = Growth.windowed.poisson,
#   Growth.windowed.SE.value = Growth.windowed.SE.poisson
#   ) #etc.

#ltlaSgeneCounts2 = ltlaSgeneCounts %>% filter(subgroup != "equivocal") %>% tsp$estimateGrowthRate(window = 14, growthRateWindow = 7)
```

```{r}
tmp = ltlaSgeneCounts2 %>% select(name,code,date,subgroup,starts_with("Growth"))

tmp2 = tmp %>% 
  filter(subgroup=="S+") %>% inner_join(
    tmp %>% filter(subgroup=="S-"),
    by = c("code","name","date"),
    suffix = c(".pos",".neg")
  ) %>% 
  mutate(
    alpha = 1-sqrt(as.numeric(maxDate-date)/as.numeric(maxDate-minDate))
  ) %>%
  mutate(
    # assume independence between S- and S+ growth rates and that CIs are normally distributes
    Growth.windowed.delta = Growth.windowed.value.pos - Growth.windowed.value.neg,
    Growth.windowed.SE.delta = sqrt(Growth.windowed.SE.value.pos^2+Growth.windowed.SE.value.neg^2),
    Growth.windowed.ProbPos.delta = 1-pnorm(0,mean=Growth.windowed.delta,sd=Growth.windowed.SE.delta),
    Growth.delta = Growth.value.pos - Growth.value.neg,
    Growth.SE.delta = sqrt(Growth.SE.value.pos^2+Growth.SE.value.neg^2),
    Growth.ProbPos.delta = 1-pnorm(0,mean=Growth.delta,sd=Growth.SE.delta)
  ) %>%
  rename(
    `S- Growth Rate` = Growth.value.neg,
    `S+ Growth Rate` = Growth.value.pos
  )

# thanks stackoverflow: https://stats.stackexchange.com/questions/30394/how-to-perform-two-sample-t-tests-in-r-by-inputting-sample-statistics-rather-tha
# m1, m2: the sample means
# s1, s2: the sample standard deviations
# n1, n2: the same sizes
# m0: the null value for the difference in means to be tested for. Default is 0. 
# equal.variance: whether or not to assume equal variance. Default is FALSE. 
differenceStat <- function(m1,m2,s1,s2,n1,n2)
{
    se <- sqrt( (s1^2/n1) + (s2^2/n2) )
    # welch-satterthwaite df
    df <- ( (s1^2/n1 + s2^2/n2)^2 )/( (s1^2/n1)^2/(n1-1) + (s2^2/n2)^2/(n2-1) )
    t <- (m1-m2)/se 
    # difference of means * (1-p-value)
    out = (m1-m2) * (1-2*pt(-abs(t),df))    
    return(out) 
}

diagonals= tibble(
  intercept = c(0.2,0.1,0,-0.1,-0.2),
  slope = 1,
  label = forcats::as_factor(c("-20%","-10%","same","+10%","+20%"))
)


maxDate = max(tmp2$date)-4
minDate = maxDate-28

topNgrowth = tmp2 %>% filter(date==maxDate) %>% mutate(
  diffStat = differenceStat(
    Growth.windowed.value.pos,
    Growth.windowed.value.neg,
    Growth.windowed.SE.value.pos,
    Growth.windowed.SE.value.neg,
    Growth.windowed.Window.value.pos,
    Growth.windowed.Window.value.neg
  )) %>%
  arrange(desc(diffStat)) %>%
  filter(row_number() <= 10) #%>%
  #select(name,code)

# topNgrowth = tmp2 %>% filter(date==maxDate) %>% mutate(
#   #diffStat = Growth.windowed.poisson.pos-Growth.windowed.poisson.neg
#     diffStat = differenceStat(
#       Growth.windowed.poisson.pos,
#       Growth.windowed.poisson.neg,
#       Growth.windowed.SE.poisson.pos,
#       Growth.windowed.SE.poisson.neg,
#       14,14
#     )
#   ) %>% 
#   ungroup() %>%
#   arrange(desc(diffStat)) %>% 
#   filter(row_number() <= 10) %>% 
#   select(name,code)

pathData = tmp2 %>% semi_join(topNgrowth, by="code") %>% 
  filter(date>minDate & date<=maxDate) %>% 
  
#pointData = tmp2 %>% filter(date==maxDate & abs(Growth.windowed.value.pos) >0.01) %>% mutate(alpha = 1)
labelData = tmp2 %>% 
  semi_join(topNgrowth, by="code") %>% 
  filter(date==maxDate) %>% 
  mutate(alpha = 1) %>%
  rename(
    `S- Growth Rate` = Growth.value.neg,
    `S+ Growth Rate` = Growth.value.pos
  )

ggplot(pathData, aes(x=`S- Growth Rate`, y=`S+ Growth Rate`, group=name, alpha=alpha))+
  coord_fixed(xlim=c(-0.2,0.2),ylim=c(-0.2,0.2))+
  geom_hline(yintercept = 0,colour="grey40")+
  geom_vline(xintercept = 0,colour="grey40")+
  geom_abline(data = diagonals, aes(intercept = intercept,slope=slope, colour=label), inherit.aes=FALSE)+
  #geom_abline(slope = 1, intercept = 0.1, colour="#FFB0B0")+
  #geom_abline(slope = 1, intercept = -0.1, colour="#B0FFB0")+
  #geom_abline(slope = 1, intercept = 0.2, colour="#FF8080")+
  #geom_abline(slope = 1, intercept = -0.2, colour="#B0FFB0")+
  #geom_point(data=pointData,colour="#B0B0FF")+
  ggrepel::geom_text_repel(data=labelData,aes(label=name),colour="grey40",min.segment.length = 0,box.padding = 2, 
                           max.overlaps = Inf)+
  geom_path(colour="blue")+
  geom_point(data=labelData,colour="blue")+
  
  guides(alpha="none")+
  xlab("S- growth rate")+
  ylab("S+ growth rate")

# ggplot(pathData, aes(x=Growth.windowed.value.neg, y=Growth.windowed.value.pos, group=name, alpha=alpha))+
#   coord_fixed(xlim=c(-0.15,0.15),ylim=c(-0.15,0.15))+
#   geom_hline(yintercept = 0,colour="grey40")+
#   geom_vline(xintercept = 0,colour="grey40")+
#   geom_abline(data = diagonals, aes(intercept = intercept,slope=slope, colour=label))+
#   #geom_abline(slope = 1, intercept = 0.1, colour="#FFB0B0")+
#   #geom_abline(slope = 1, intercept = -0.1, colour="#B0FFB0")+
#   #geom_abline(slope = 1, intercept = 0.2, colour="#FF8080")+
#   #geom_abline(slope = 1, intercept = -0.2, colour="#B0FFB0")+
#   #geom_point(data=pointData,colour="#B0B0FF")+
#   ggrepel::geom_text_repel(data=labelData,aes(label=name),colour="grey40",min.segment.length = 0,box.padding = 2, 
#                            max.overlaps = Inf)+
#   geom_path(colour="blue")+
#   geom_point(data=labelData,colour="blue")+
#   
#   guides(alpha="none")+
#   xlab("S- growth rate")+
#   ylab("S+ growth rate")

topNgrowth = tmp2 %>% filter(date==maxDate) %>%
  arrange(desc(Growth.windowed.ProbPos.delta*Growth.windowed.delta)) %>%
  filter(row_number() <= 10) %>%
  select(name,code)

(ggplot(tmp2 %>% filter(date>"2020-10-01"),
       aes(
         x=date,
         y=Growth.windowed.delta,
         group=name#,
         #colour=Growth.windowed.ProbPos.delta
         )
        )+geom_line(alpha=0.2)+geom_hline(yintercept = 0,colour="red"))  %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/AllGrowthDelta"))

(ggplot(tmp2 %>% semi_join(topNgrowth) %>% filter(date>"2020-02-01"),
       aes(x=date,
           y=Growth.windowed.delta,
           ymin=Growth.windowed.delta-1.96*Growth.windowed.SE.delta,
           ymax=Growth.windowed.delta+1.96*Growth.windowed.SE.delta,
           group=name)
       )+
  geom_ribbon(alpha=0.2,fill="grey50")+
  geom_line(mapping=aes(colour=name)))  %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/Top10GrowthDelta"))

rollingPercent = allVoc %>% #combinedSpositives %>% 
    filter(!is.na(type)) %>%
    group_by(LTLA_name, type, date=specimen_date) %>%
    summarise(count = n()) %>%
    ungroup() %>%
    tidyr::complete(LTLA_name,type,date,fill=list(count=0)) %>%
    group_by(LTLA_name, type) %>% 
    arrange(date) %>%
    mutate(Roll.count = stats::filter(count,filter=rep(1,28)/28,sides=1)) %>%
    filter(!is.na(Roll.count)) %>% 
    group_by(LTLA_name,date) %>% 
    mutate(binom::binom.confint(Roll.count, sum(Roll.count), conf.level = 0.95, methods = "wilson")) %>% 
    mutate(weight = sum(Roll.count)) %>% 
    ungroup()

growthVPercent = tmp2 %>% inner_join(rollingPercent, by = c("name"="LTLA_name","date"="date")) %>%
  mutate(growthSigHigher = ifelse(Growth.windowed.ProbPos.delta>0.0975,"sig","ns"))

(ggplot(growthVPercent %>% filter(date==max(date)), aes(x=mean,y=Growth.windowed.delta,size=weight))+geom_point(alpha=0.3)+facet_wrap(vars(type),scales="free")+
  #geom_smooth(method = "lm",fullrange=TRUE)+
  coord_cartesian(xlim = c(0,1))) %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/SGeneGrowthDeltaVsPercentSequencing"))

(ggplot(growthVPercent %>% filter(date==max(date)), aes(x=Roll.count,y=Growth.windowed.delta,size=weight))+geom_point(alpha=0.3)+facet_wrap(vars(type))
  #geom_smooth(method = "lm",fullrange=TRUE)+
  ) %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/SGeneGrowthDeltaVsCountSequencing"))

ggplot(growthVPercent %>% filter(type %in% c("B.1.351","B.1.617.1","B.1.617.2","P.1 & P.2") & date==maxDate), aes(x=type,colour=growthSigHigher,y=mean))+geom_boxplot()+scale_y_log10()

```


```{r}
# map = dpc$geog$getMap("LAD19")
# 
# map2 = map %>% inner_join(rollingPercent, by=c("name"="LTLA_name"))
# ggplot(map2 %>% filter(date==maxDate))+geom_sf(aes(fill=mean))+facet_wrap(vars(type))+scale_fill_viridis_c()
#                    
```

```{r}

map3 = map %>% inner_join(tmp2, by="code")
(ggplot(map3 %>% filter(date==maxDate))+geom_sf(aes(fill=Growth.windowed.delta, colour=ifelse(Growth.windowed.ProbPos.delta>0.0975,"sig","ns")),size=0.3)+scale_color_manual(values = c("sig"="white","ns"="black"))+scale_fill_gradient2(high = "yellow", mid="black",low="cyan",limits=c(-0.15,0.15),oob=scales::squish)) %>%
    standardPrintOutput::saveFullPageFigure(paste0("~/Dropbox/covid19/sa-variant/SgeneGrowthDelta-England"))

lapply(names(locations), function(name) {
  limits = locations[[name]]
  xlim = c(limits[2],limits[4])
  ylim = c(limits[1],limits[3])
  (ggplot(map3 %>% filter(date==max(date)))+geom_sf(aes(fill=Growth.windowed.delta))+scale_fill_gradient2(high = "orange",low="blue",limits=c(-0.15,0.15),oob=scales::squish)+coord_sf(xlim=xlim,ylim=ylim)) %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/SgeneGrowthDelta-",name))
})

```

```{r}
topNgrowth %>% inner_join(rollingPercent %>% filter(date==maxDate) %>% mutate(percent =sprintf("%1.1f%%",mean*100)) %>% select(LTLA_name,type,percent)  %>% pivot_wider(values_from = percent, names_from=type), by = c("name"="LTLA_name"))
```


```{r}
ltlaSgeneCounts3 = ltlaSgeneCounts2 %>% dpc$demog$findDemographics()
ltlaSgeneCounts3 = ltlaSgeneCounts3 %>% inner_join(rollingPercent %>% rename(variant=type), by = c("name"="LTLA_name","date"="date"))


tsp$plotGrowthIncidence(ltlaSgeneCounts3 %>% filter(variant == "B.1.617.2") %>% group_by(name, variant,subgroup),plotDates = maxDate,timespan = 28,colour = mean*weight,populationExpr = population,maxSize = 0.6)+facet_wrap(vars(variant,subgroup))+scale_colour_viridis_b()

```


```{r}

sPosll = ll2 %>% select(-specimen_date) %>% inner_join(sgPosUniq %>% filter(sgtf_under30CT == 0) %>% select(FINALID,specimen_date), by="FINALID")

sPosIncid = sPosll %>% dpc$spim$getLineListIncidence(ageBreaks = c(18,35,65,75)) %>% tsp$aggregateGender() %>% dpc$demog$findDemographics()
tsp$plotIncidenceQuantiles(covidTimeseries = sPosIncid %>% filter(name=="England"), dates = "2021-02-01", denominatorExpr = population/1000000, colour=ageCat)+scale_y_continuous(trans="log1p")

tsp$plotRt(covidRtTimeseries = sPosIncid %>% filter(name=="England"), dates = "2021-02-01", colour=ageCat,rtlim = c(0.25,1.5))
```

```{r}
notSNegll = ll %>% semi_join(sgll %>% filter(is.na(sgtf_under30CT) | sgtf_under30CT == 1), by="FINALID")
sNotNegIncid = notSNegll %>% dpc$spim$getLineListIncidence() %>% tsp$aggregateGender() %>% dpc$demog$findDemographics()
tsp$plotIncidenceQuantiles(covidTimeseries = sNotNegIncid, dates = "2021-02-01", denominatorExpr = population/1000000)+facet_wrap(vars(name))+scale_y_continuous(trans="log1p")

tsp$plotRt(covidRtTimeseries = sNotNegIncid, dates = "2021-02-01",rtlim = c(0.25,1.5))+facet_wrap(vars(name))

```

```{r}
sgUniq = sgll %>% group_by(FINALID) %>% 
  filter(CDR_Specimen_Request_SK==min(CDR_Specimen_Request_SK,na.rm=TRUE)) 
llWSgene = ll2 %>% filter(asymptomatic_indicator != "Y") %>% inner_join(sgll %>% filter(!is.na(sgtf_under30CT)) %>% mutate(sGene = ifelse(sgtf_under30CT == 1,"S neg","S pos")) %>% select(FINALID, sGene), by="FINALID")
sGeneIncid = llWSgene %>% dpc$spim$getLineListIncidence(subgroup = sGene) %>% dpc$demog$findDemographics()
tsp$plotIncidenceQuantiles(covidTimeseries = sGeneIncid, dates = "2021-02-01", colour = subgroup, denominatorExpr = population/1000000)+facet_wrap(vars(name))+scale_y_continuous(trans="log1p")

sRT14 = sGeneIncid %>% filter(name != "Unknown (England)") %>% tsp$estimateRt(quick = TRUE,window = 14)
tsp$plotRt(covidRtTimeseries = sRT14, dates = "2021-02-01",rtlim = c(0.25,1.5), colour = subgroup)+facet_wrap(vars(name))

```


# Bolton cluster

```{r}

# sgeneIncidence = sgll %>% select(FINALID,specimen_date,sgtf_under30CT) %>% left_join(ll2 %>% select(-specimen_date), by=c("FINALID"))
# sgeneIncidence2 = sgeneIncidence %>% filter(LTLA_code %in% areasOfConcern$`Bolton etc`)
# #sgeneIncidence2 %>% group_by(FINALID) %>% filter(n() > 1) %>% arrange(specimen_date) %>% filter(lag(specimen_date)<specimen_date-56) %>% View()
# sgeneIncidence3 = sgeneIncidence2 %>% group_by(FINALID) %>% filter(specimen_date == min(specimen_date)) %>% ungroup()
sgSymptIncid = sgllLoc %>% filter(LTLA_code %in% areasOfConcern$`NW Cluster`) %>% dpc$spim$getLineListIncidence(ageBreaks = c(6,18,45,55,75),subgroup = sgtf_under30CT,filterExpr = asymptomatic_indicator=="N",codeTypes = "LAD") 
#sgSymptIncid = sgllLoc %>% filter(LTLA_code %in% areasOfConcern$`Bolton etc`) %>% dpc$spim$getLineListIncidence(ageBreaks = c(6,18,45,55,75),subgroup = sgtf_under30CT,codeTypes = "LAD") 
tmp = sgSymptIncid %>% dpc$demog$findDemographics()
tmp2 = tmp %>% mutate(code="Bolton etc", name="Bolton etc") %>% ukcovidtools::covidStandardDateGrouping() %>% summarise(value=sum(value), population = sum(population))
tmp2 = tmp2 %>% mutate(subgroup=case_when(
  subgroup==0 ~ "S+",
  subgroup==1 ~ "S-",
  TRUE ~ "equivocal"
))
(tsp$plotIncidenceQuantiles(tmp2 %>% filter(date>"2021-03-01" & ageCat !="unknown"),colour = subgroup,denominatorExpr = population/1000000)+facet_wrap(vars(ageCat))+ylab("daily incidence per 1M")) %>%
  standardPrintOutput::saveThirdPageFigure("~/Dropbox/covid19/sa-variant/BoltonIncidencePer1M")
tmp3 = tmp2 %>% tsp$aggregateAge() %>% tsp$estimateRt(window = 14,quick = FALSE)
tsp$plotRt(tmp3,colour = subgroup,dates =  "2021-03-01", window=14, rtlim = c(0,3)) %>%
  standardPrintOutput::saveThirdPageFigure("~/Dropbox/covid19/sa-variant/BoltonRt")

ggplot(
  sgllLoc %>% #filter(LTLA_code %in% areasOfConcern$`NW Cluster`) %>% 
    filter(specimen_date>"2021-04-01") %>% mutate(sgene=case_when(
  sgtf_under30CT==0 ~ "S+",
  sgtf_under30CT==1 ~ "S-",
  TRUE ~ "equivocal"
)) ,aes(x=age, colour=sgene))+geom_histogram(binwidth = 1)+facet_wrap(vars(sgene), scales = "free")

```
```{r}



```

```{r}

ltlaSgeneCountsAll = sgllLoc %>% dpc$spim$getLineListIncidence(ageBreaks = NULL,codeTypes = "LAD", subgroup = sGene)

ltlaToAoc = function(l) l %>% 
  dpc$demog$findDemographics() %>% 
  inner_join(aocDf, by = c("code"="ltlaCode")) %>% 
  select(-code,-name,-name.original) %>%
  rename(name = area) %>%
  mutate(
    name = as.character(name),
    code = name,
    codeType = "AOC"
  ) %>%
  ukcovidtools::covidStandardDateGrouping() %>%
  summarise(value = sum(value), population = sum(population)) %>%
  ungroup()

aocSgeneCounts = bind_rows(
  ltlaToAoc(ltlaSgeneCounts),
  symptomaticPillar2CtrySgeneCounts %>% dpc$demog$findDemographics()
)

aocSgeneCounts %>% filter(subgroup!="equivocal") %>% readr::write_csv("~/Dropbox/covid19/sa-variant/symptomatic-pillar2-cases-by-aoc-and-sgene.csv")

aocSgeneCountsAll = bind_rows(
  ltlaToAoc(ltlaSgeneCountsAll),
  allPillar2CtrySgeneCounts %>% dpc$demog$findDemographics()
)

aocSgeneCountsAll %>% filter(subgroup!="equivocal") %>% readr::write_csv("~/Dropbox/covid19/sa-variant/all-pillar2-cases-by-aoc-and-sgene.csv")

aocDf %>% readr::write_csv("~/Dropbox/covid19/sa-variant/ltla-to-aoc.csv")


  
tsp$plotIncidenceQuantiles(aocSgeneCounts, colour=subgroup, dates="2021-03-01")+facet_wrap(vars(name))+scale_y_continuous(trans="log1p", breaks=c(1,2,5,10,20,50,100,200,500,1000,2000,5000,10000))
tsp$plotIncidenceQuantiles(aocSgeneCountsAll, colour=subgroup, dates="2021-03-01")+facet_wrap(vars(name))+scale_y_continuous(trans="log1p", breaks=c(1,2,5,10,20,50,100,200,500,1000,2000,5000,10000))



clusters = list(
  "CommunityIndia2" = c("NW Cluster","Bedford etc","Sefton & Lpl"),
  "MixedIndia2" = c("Leicester","Nottingham","West London"),
  "MixedSA" = c("East London"),
  "Reference" = c("England")
)

aocSgeneCountsGrowth = aocSgeneCounts %>% tsp$estimateGrowthRate(window = 56,growthRateWindow = 7)

lapply(names(clusters),function(clustName) {
  (
    tsp$plotGrowthRate(
      aocSgeneCountsGrowth %>% filter(name %in% clusters[[clustName]]) %>% filter(subgroup!="equivocal"),
      colour = subgroup,dates="2021-03-01",
      rlim=c(-0.25,0.25),
      growthVar = Growth.poisson,
      growthSEVar = Growth.SE.poisson,
    )+facet_wrap(vars(name))+standardPrintOutput::watermark(lab="provisional")+ylab("r - sliding 8 week window")
  ) %>%   
  standardPrintOutput::saveThirdPageFigure(paste0("~/Dropbox/covid19/sa-variant/sGeneAOCHighGrowth-",clustName))
})

aocSgeneCountsRt = aocSgeneCounts %>% tsp$estimateRt(window = 14)
lapply(names(clusters),function(clustName) {
  (
    tsp$plotRt(
      aocSgeneCountsRt %>% filter(name %in% clusters[[clustName]]) %>% filter(subgroup!="equivocal"),
      colour = subgroup,
      dates="2021-03-01",
      rtlim=c(0,3.5)
    )+
    facet_wrap(vars(name))+
    standardPrintOutput::watermark(lab="provisional")) %>%   
  standardPrintOutput::saveThirdPageFigure(paste0("~/Dropbox/covid19/sa-variant/sGeneAOCHighRt-",clustName))
})




# aocSgeneCountsLogGrowth = aocSgeneCounts %>% tsp$logIncidenceStats(smoothingWindow = 21,growthRateWindow = 7, leftSided=TRUE)
# (tsp$plotGrowthRate(aocSgeneCountsLogGrowth %>% 
#               filter(name %in% c("NW Cluster","Bedford etc","Sefton & Lpl")) %>% 
#               filter(subgroup!="equivocal"),colour = subgroup,dates="2021-03-01",rlim=c(-0.25,0.25))+facet_wrap(vars(name))+standardPrintOutput::watermark(lab="provisional")) %>%   
#   standardPrintOutput::saveThirdPageFigure("~/Dropbox/covid19/sa-variant/sGeneAOCHighGrowthLog")
# 
# 
# aocSgeneCountsRtAll = aocSgeneCountsAll %>% tsp$estimateRt(window = 14)
# (tsp$plotRt(aocSgeneCountsRt %>% filter(subgroup!="equivocal"),colour = subgroup,dates="2021-03-01",rtlim=c(0,3.5))+facet_wrap(vars(name))+standardPrintOutput::watermark(lab="provisional")) %>%   
#   standardPrintOutput::saveHalfPageFigure("~/Dropbox/covid19/sa-variant/sGeneAllAOCRt")


#ggplot(ltlaSgeneCounts, aes(x=date,y=value))
# ltlasSgeneCounts is asymptomatic == "N"
#ltlaSgeneCounts2 = ltlaSgeneCounts %>% filter(subgroup != "equivocal") %>% tsp$logIncidenceStats(growthRateWindow = 7,nocache=TRUE)
ltlaSgeneCounts2 = ltlaSgeneCounts %>% filter(subgroup != "equivocal") %>% tsp$estimateGrowthRate(window = 14, growthRateWindow = 7)
# ltlaSgeneCounts2 = ltlaSgeneCounts2 %>% rename(
#   Growth.windowed.value = Growth.windowed.poisson,
#   Growth.windowed.SE.value = Growth.windowed.SE.poisson
#   ) #etc.

#ltlaSgeneCounts2 = ltlaSgeneCounts %>% filter(subgroup != "equivocal") %>% tsp$estimateGrowthRate(window = 14, growthRateWindow = 7)
```

```{r}
tmp = ltlaSgeneCounts2 %>% select(name,code,date,subgroup,starts_with("Growth"))

tmp2 = tmp %>% 
  filter(subgroup=="S+") %>% inner_join(
    tmp %>% filter(subgroup=="S-"),
    by = c("code","name","date"),
    suffix = c(".pos",".neg")
  ) %>% 
  mutate(
    alpha = 1-sqrt(as.numeric(maxDate-date)/as.numeric(maxDate-minDate))
  ) %>%
  mutate(
    # assume independence between S- and S+ growth rates and that CIs are normally distributes
    Growth.windowed.delta = Growth.windowed.value.pos - Growth.windowed.value.neg,
    Growth.windowed.SE.delta = sqrt(Growth.windowed.SE.value.pos^2+Growth.windowed.SE.value.neg^2),
    Growth.windowed.ProbPos.delta = 1-pnorm(0,mean=Growth.windowed.delta,sd=Growth.windowed.SE.delta),
    Growth.delta = Growth.value.pos - Growth.value.neg,
    Growth.SE.delta = sqrt(Growth.SE.value.pos^2+Growth.SE.value.neg^2),
    Growth.ProbPos.delta = 1-pnorm(0,mean=Growth.delta,sd=Growth.SE.delta)
  ) %>%
  rename(
    `S- Growth Rate` = Growth.value.neg,
    `S+ Growth Rate` = Growth.value.pos
  )

# thanks stackoverflow: https://stats.stackexchange.com/questions/30394/how-to-perform-two-sample-t-tests-in-r-by-inputting-sample-statistics-rather-tha
# m1, m2: the sample means
# s1, s2: the sample standard deviations
# n1, n2: the same sizes
# m0: the null value for the difference in means to be tested for. Default is 0. 
# equal.variance: whether or not to assume equal variance. Default is FALSE. 
differenceStat <- function(m1,m2,s1,s2,n1,n2)
{
    se <- sqrt( (s1^2/n1) + (s2^2/n2) )
    # welch-satterthwaite df
    df <- ( (s1^2/n1 + s2^2/n2)^2 )/( (s1^2/n1)^2/(n1-1) + (s2^2/n2)^2/(n2-1) )
    t <- (m1-m2)/se 
    # difference of means * (1-p-value)
    out = (m1-m2) * (1-2*pt(-abs(t),df))    
    return(out) 
}

diagonals= tibble(
  intercept = c(0.2,0.1,0,-0.1,-0.2),
  slope = 1,
  label = forcats::as_factor(c("-20%","-10%","same","+10%","+20%"))
)


maxDate = max(tmp2$date)-4
minDate = maxDate-28

topNgrowth = tmp2 %>% filter(date==maxDate) %>% mutate(
  diffStat = differenceStat(
    Growth.windowed.value.pos,
    Growth.windowed.value.neg,
    Growth.windowed.SE.value.pos,
    Growth.windowed.SE.value.neg,
    Growth.windowed.Window.value.pos,
    Growth.windowed.Window.value.neg
  )) %>%
  arrange(desc(diffStat)) %>%
  filter(row_number() <= 10) #%>%
  #select(name,code)

# topNgrowth = tmp2 %>% filter(date==maxDate) %>% mutate(
#   #diffStat = Growth.windowed.poisson.pos-Growth.windowed.poisson.neg
#     diffStat = differenceStat(
#       Growth.windowed.poisson.pos,
#       Growth.windowed.poisson.neg,
#       Growth.windowed.SE.poisson.pos,
#       Growth.windowed.SE.poisson.neg,
#       14,14
#     )
#   ) %>% 
#   ungroup() %>%
#   arrange(desc(diffStat)) %>% 
#   filter(row_number() <= 10) %>% 
#   select(name,code)

pathData = tmp2 %>% semi_join(topNgrowth, by="code") %>% 
  filter(date>minDate & date<=maxDate) %>% 
  
#pointData = tmp2 %>% filter(date==maxDate & abs(Growth.windowed.value.pos) >0.01) %>% mutate(alpha = 1)
labelData = tmp2 %>% 
  semi_join(topNgrowth, by="code") %>% 
  filter(date==maxDate) %>% 
  mutate(alpha = 1) %>%
  rename(
    `S- Growth Rate` = Growth.value.neg,
    `S+ Growth Rate` = Growth.value.pos
  )

ggplot(pathData, aes(x=`S- Growth Rate`, y=`S+ Growth Rate`, group=name, alpha=alpha))+
  coord_fixed(xlim=c(-0.2,0.2),ylim=c(-0.2,0.2))+
  geom_hline(yintercept = 0,colour="grey40")+
  geom_vline(xintercept = 0,colour="grey40")+
  geom_abline(data = diagonals, aes(intercept = intercept,slope=slope, colour=label), inherit.aes=FALSE)+
  #geom_abline(slope = 1, intercept = 0.1, colour="#FFB0B0")+
  #geom_abline(slope = 1, intercept = -0.1, colour="#B0FFB0")+
  #geom_abline(slope = 1, intercept = 0.2, colour="#FF8080")+
  #geom_abline(slope = 1, intercept = -0.2, colour="#B0FFB0")+
  #geom_point(data=pointData,colour="#B0B0FF")+
  ggrepel::geom_text_repel(data=labelData,aes(label=name),colour="grey40",min.segment.length = 0,box.padding = 2, 
                           max.overlaps = Inf)+
  geom_path(colour="blue")+
  geom_point(data=labelData,colour="blue")+
  
  guides(alpha="none")+
  xlab("S- growth rate")+
  ylab("S+ growth rate")

# ggplot(pathData, aes(x=Growth.windowed.value.neg, y=Growth.windowed.value.pos, group=name, alpha=alpha))+
#   coord_fixed(xlim=c(-0.15,0.15),ylim=c(-0.15,0.15))+
#   geom_hline(yintercept = 0,colour="grey40")+
#   geom_vline(xintercept = 0,colour="grey40")+
#   geom_abline(data = diagonals, aes(intercept = intercept,slope=slope, colour=label))+
#   #geom_abline(slope = 1, intercept = 0.1, colour="#FFB0B0")+
#   #geom_abline(slope = 1, intercept = -0.1, colour="#B0FFB0")+
#   #geom_abline(slope = 1, intercept = 0.2, colour="#FF8080")+
#   #geom_abline(slope = 1, intercept = -0.2, colour="#B0FFB0")+
#   #geom_point(data=pointData,colour="#B0B0FF")+
#   ggrepel::geom_text_repel(data=labelData,aes(label=name),colour="grey40",min.segment.length = 0,box.padding = 2, 
#                            max.overlaps = Inf)+
#   geom_path(colour="blue")+
#   geom_point(data=labelData,colour="blue")+
#   
#   guides(alpha="none")+
#   xlab("S- growth rate")+
#   ylab("S+ growth rate")

topNgrowth = tmp2 %>% filter(date==maxDate) %>%
  arrange(desc(Growth.windowed.ProbPos.delta*Growth.windowed.delta)) %>%
  filter(row_number() <= 10) %>%
  select(name,code)

(ggplot(tmp2 %>% filter(date>"2020-10-01"),
       aes(
         x=date,
         y=Growth.windowed.delta,
         group=name#,
         #colour=Growth.windowed.ProbPos.delta
         )
        )+geom_line(alpha=0.2)+geom_hline(yintercept = 0,colour="red"))  %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/AllGrowthDelta"))

(ggplot(tmp2 %>% semi_join(topNgrowth) %>% filter(date>"2020-02-01"),
       aes(x=date,
           y=Growth.windowed.delta,
           ymin=Growth.windowed.delta-1.96*Growth.windowed.SE.delta,
           ymax=Growth.windowed.delta+1.96*Growth.windowed.SE.delta,
           group=name)
       )+
  geom_ribbon(alpha=0.2,fill="grey50")+
  geom_line(mapping=aes(colour=name)))  %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/Top10GrowthDelta"))

rollingPercent = allVoc %>% #combinedSpositives %>% 
    filter(!is.na(type)) %>%
    group_by(LTLA_name, type, date=specimen_date) %>%
    summarise(count = n()) %>%
    ungroup() %>%
    tidyr::complete(LTLA_name,type,date,fill=list(count=0)) %>%
    group_by(LTLA_name, type) %>% 
    arrange(date) %>%
    mutate(Roll.count = stats::filter(count,filter=rep(1,28)/28,sides=1)) %>%
    filter(!is.na(Roll.count)) %>% 
    group_by(LTLA_name,date) %>% 
    mutate(binom::binom.confint(Roll.count, sum(Roll.count), conf.level = 0.95, methods = "wilson")) %>% 
    mutate(weight = sum(Roll.count)) %>% 
    ungroup()

growthVPercent = tmp2 %>% inner_join(rollingPercent, by = c("name"="LTLA_name","date"="date")) %>%
  mutate(growthSigHigher = ifelse(Growth.windowed.ProbPos.delta>0.0975,"sig","ns"))

(ggplot(growthVPercent %>% filter(date==max(date)), aes(x=mean,y=Growth.windowed.delta,size=weight))+geom_point(alpha=0.3)+facet_wrap(vars(type),scales="free")+
  #geom_smooth(method = "lm",fullrange=TRUE)+
  coord_cartesian(xlim = c(0,1))) %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/SGeneGrowthDeltaVsPercentSequencing"))

(ggplot(growthVPercent %>% filter(date==max(date)), aes(x=Roll.count,y=Growth.windowed.delta,size=weight))+geom_point(alpha=0.3)+facet_wrap(vars(type))
  #geom_smooth(method = "lm",fullrange=TRUE)+
  ) %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/SGeneGrowthDeltaVsCountSequencing"))

ggplot(growthVPercent %>% filter(type %in% c("B.1.351","B.1.617.1","B.1.617.2","P.1 & P.2") & date==maxDate), aes(x=type,colour=growthSigHigher,y=mean))+geom_boxplot()+scale_y_log10()

```


```{r}
# map = dpc$geog$getMap("LAD19")
# 
# map2 = map %>% inner_join(rollingPercent, by=c("name"="LTLA_name"))
# ggplot(map2 %>% filter(date==maxDate))+geom_sf(aes(fill=mean))+facet_wrap(vars(type))+scale_fill_viridis_c()
#                    
```

```{r}

map3 = map %>% inner_join(tmp2, by="code")
(ggplot(map3 %>% filter(date==maxDate))+geom_sf(aes(fill=Growth.windowed.delta, colour=ifelse(Growth.windowed.ProbPos.delta>0.0975,"sig","ns")),size=0.3)+scale_color_manual(values = c("sig"="white","ns"="black"))+scale_fill_gradient2(high = "yellow", mid="black",low="cyan",limits=c(-0.15,0.15),oob=scales::squish)) %>%
    standardPrintOutput::saveFullPageFigure(paste0("~/Dropbox/covid19/sa-variant/SgeneGrowthDelta-England"))

lapply(names(locations), function(name) {
  limits = locations[[name]]
  xlim = c(limits[2],limits[4])
  ylim = c(limits[1],limits[3])
  (ggplot(map3 %>% filter(date==max(date)))+geom_sf(aes(fill=Growth.windowed.delta))+scale_fill_gradient2(high = "orange",low="blue",limits=c(-0.15,0.15),oob=scales::squish)+coord_sf(xlim=xlim,ylim=ylim)) %>%
    standardPrintOutput::saveHalfPageFigure(paste0("~/Dropbox/covid19/sa-variant/SgeneGrowthDelta-",name))
})

```

```{r}
topNgrowth %>% inner_join(rollingPercent %>% filter(date==maxDate) %>% mutate(percent =sprintf("%1.1f%%",mean*100)) %>% select(LTLA_name,type,percent)  %>% pivot_wider(values_from = percent, names_from=type), by = c("name"="LTLA_name"))
```


```{r}
ltlaSgeneCounts3 = ltlaSgeneCounts2 %>% dpc$demog$findDemographics()
ltlaSgeneCounts3 = ltlaSgeneCounts3 %>% inner_join(rollingPercent %>% rename(variant=type), by = c("name"="LTLA_name","date"="date"))


tsp$plotGrowthIncidence(ltlaSgeneCounts3 %>% filter(variant == "B.1.617.2") %>% group_by(name, variant,subgroup),plotDates = maxDate,timespan = 28,colour = mean*weight,populationExpr = population,maxSize = 0.6)+facet_wrap(vars(variant,subgroup))+scale_colour_viridis_b()

```